{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "b5e82a2c",
   "metadata": {},
   "source": [
    "# Développement d’une solution de mapping de données de séquençage à haut-débit sur un génome de référence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ba5f51b",
   "metadata": {},
   "source": [
    "### Mise en place d'un algorithme permettant de chercher un mot de longueur fixe dans un texte"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0a91cb8",
   "metadata": {},
   "source": [
    "### Plan du notebook:\n",
    "1/ Prise en main de Biopython  \n",
    "2/ Implémentation d'un algorithme \"Difference Cover size 3\" pour construire une table de suffixe à partir d'un génome en un temps linéaire  \n",
    "3/ Trouver la transformée de Burrows-Wheeler   \n",
    "4/ Recherche d'un mot dans un texte grâce à la transformée de Burrows-Wheeler  \n",
    "5/ Importation des données expérimentales     \n",
    "6/ Créer et aligner les kmers de nos reads  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fe6f1e10",
   "metadata": {},
   "source": [
    "## 1/ Biopython"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38e8c197",
   "metadata": {},
   "source": [
    "Importation des fichiers, tests des informations générales des fichiers."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "71bfc888",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Defaulting to user installation because normal site-packages is not writeable\n",
      "Requirement already satisfied: pysam in /home/azarkua/.local/lib/python3.10/site-packages (0.22.0)\n",
      "\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.2.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.3.1\u001b[0m\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpython3 -m pip install --upgrade pip\u001b[0m\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "63868023",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from collections import Counter\n",
    "import time\n",
    "import matplotlib.pyplot as plt\n",
    "import pstats\n",
    "import cProfile\n",
    "import copy\n",
    "import math\n",
    "import pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6149957a",
   "metadata": {},
   "outputs": [],
   "source": [
    "path_to_genome_file=\"/home/azarkua/Documents/2023-2024/omiques2/developement/omique2/genome.fna\"\n",
    "path_to_reads_file=\"/home/azarkua/Documents/2023-2024/omiques2/developement/omique2/reads.fq\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0bdf54b0",
   "metadata": {},
   "source": [
    "path_to_genome_file=\"/home/mvernier/Documents/2023-2024/omique2/projet/GCF_000002765.5_GCA_000002765_genomic.fna\"\n",
    "path_to_reads_file=\"/home/mvernier/Documents/2023-2024/omique2/projet/single_Pfal_dat.fq\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ef73a1e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome=[]\n",
    "nucleotide_genome=0\n",
    "for seq_record in SeqIO.parse(path_to_genome_file, \"fasta\"):\n",
    "    genome.append(seq_record)\n",
    "    nucleotide_genome+=len(seq_record.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "aef5a3e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Informations contenues dans un élément de 'genome'.\n",
      "ID: NC_004325.2\n",
      "Name: NC_004325.2\n",
      "Description: NC_004325.2 Plasmodium falciparum 3D7 genome assembly, chromosome: 1\n",
      "Number of features: 0\n",
      "Seq('TGAACCCTaaaacctaaaccctaaaccctaaaccctgaaccctaaaccctgaac...agg')\n"
     ]
    }
   ],
   "source": [
    "print(\"Informations contenues dans un élément de 'genome'.\")\n",
    "print(genome[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b2880ec2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Notre génome contient 15 séquences (14 chromosomes et 1 plasmide) et est constitué de 23326872 nucléotides.\n"
     ]
    }
   ],
   "source": [
    "print(\"Notre génome contient \"+str(len(genome))+\" séquences (\"+ str(len(genome)-1) +\" chromosomes et 1 plasmide) et est constitué de \"+str(nucleotide_genome)+\" nucléotides.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "92d6fa3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "reads=[]\n",
    "nucleotide_reads=0\n",
    "for seq_record in SeqIO.parse(path_to_reads_file, \"fastq\"):\n",
    "    reads.append(seq_record)\n",
    "    nucleotide_reads+=len(seq_record.seq)\n",
    "    #if len(seq_record.seq)!= 100:\n",
    "    #    print(\"False\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8950178e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Informations contenues dans un élément de 'reads'.\n",
      "ID: NC_004325.2-99990\n",
      "Name: NC_004325.2-99990\n",
      "Description: NC_004325.2-99990\n",
      "Number of features: 0\n",
      "Per letter annotation for: phred_quality\n",
      "Seq('TCTTTTAACATACCTAAGAAGGAATACATTTTACACTTTACCTATTATTTATTC...ATA')\n"
     ]
    }
   ],
   "source": [
    "print(\"Informations contenues dans un élément de 'reads'.\")\n",
    "print(reads[10])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "daf7f246",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Nous avons 1500000 reads qui contiennent 150000000 nucléotides.\n",
      "Chaque read a une longueur de 100 nucléotide.\n"
     ]
    }
   ],
   "source": [
    "print(\"Nous avons \"+str(len(reads))+\" reads qui contiennent \"+str(nucleotide_reads)+\" nucléotides.\")\n",
    "print(\"Chaque read a une longueur de \"+str(int(nucleotide_reads/len(reads)))+\" nucléotide.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "51210559",
   "metadata": {},
   "source": [
    "## 2/ DC3\n",
    "### Première étape de DC3: division des positions de notre génome, tri des positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7166cabb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def asciiDC3 (seq) : \n",
    "    \"\"\"\n",
    "    Create sequence of ascii equivalent of each element of given list parameter.\n",
    "    Add three sentinel numbers necessary for DC3 algorithm.\n",
    "    \n",
    "    Args:\n",
    "        seq (list of str): list of elements to transform\n",
    "    \n",
    "    Return:\n",
    "        (list of int)\n",
    "    \"\"\"\n",
    "    asc=[]\n",
    "    for i in seq : \n",
    "        asc.append(ord(i))\n",
    "    \n",
    "    return asc+[0,0,0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13dc9fd0",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "Posons n, la taille de notre séquence d'ADN. Ainsi, en parlant d'une complexité de $O(n)$, on a une complexité linéaire selon la taille de la séquence fournie.\n",
    "Parcourir la séquence 1 fois implique que asciiDC3 a une compléxité $O(n)$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "c807cbc7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def position1_2 (asc):\n",
    "    \"\"\"\n",
    "    Create list of positions not multiple of 3\n",
    "    \n",
    "    Args:\n",
    "        asc (list of int): sequence we want to extract a suffix array from\n",
    "    \n",
    "    Return:\n",
    "         ind1+ind2 (list of int): list of positions not multiple of 3\n",
    "    \"\"\"\n",
    "    ind1=[]\n",
    "    ind2=[]\n",
    "    for k in range(len(asc)-2):\n",
    "        #attention on a peut etre fait de la merde ici, pas sure du -2\n",
    "        if k%3==1 :\n",
    "            ind1.append(k)\n",
    "        if k%3==2:\n",
    "            ind2.append(k)\n",
    "    # dans le cas où T a un nombre d'élément multiple de 3, r12 ne contiendra pas le triplet [0,0,0].\n",
    "    # Or, la fonction, \"removesentinel\" retire 1 élément notre suffix table, pour justement se débarrasser de [0,0,0]\n",
    "    # J'ai donc décider de rajouter artificiellement [0,0,0] dans les cas où il n'est pas déjà là.\n",
    "    if len(asc)%3==0:\n",
    "        ind1.append(len(asc)-1-2)\n",
    "        \n",
    "    return ind1+ind2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1277ec13",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "Parcourir la séquence 1 fois implique que position1_2 a une compléxité $O(n)$, n étant la longueur de la séquence "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "983e5cfc",
   "metadata": {},
   "outputs": [],
   "source": [
    "def radix_with_p12(p,t):\n",
    "    \"\"\"\n",
    "    Creates a list of couples containing:\n",
    "    - sequences of three elements from \"t\" staring from positions in parameter \"p\"\n",
    "    - the position p\n",
    "    \n",
    "    Args : \n",
    "        p (list of int): list of positions not multiple of 3 \n",
    "        t (list of int): sequence we want to extract a suffix array from\n",
    "    \n",
    "    Return : \n",
    "        r (list of list): list of triplets and their starting positions\n",
    "    \n",
    "    \"\"\"\n",
    "    r=[]\n",
    "    for i in range(len(p)):\n",
    "        index=p[i]\n",
    "        r.append([[t[index],t[index+1], t[index+2]],index])\n",
    "    return r"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a021a4e8",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "Parcourir la séquence 1 fois implique que radix_with_p12 a une compléxité $O(n)$, n étant la longueur de la séquence "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "862523e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sort_with_p12(array, alphabet, column_number):\n",
    "    \"\"\"\n",
    "    Sorts parameter \"array\" using Radix Sort\n",
    "    \n",
    "    Args:\n",
    "        array (list of list): list of triplets and their positions.\n",
    "        \n",
    "        alphabet (dictionnary): alphabet of our array. Ouput of function \"alphabetT\" or \"alphabetR0_with_p12\"\n",
    "        \n",
    "        columnNumber (int): number of last column of the lists in our paramater \"array\". \n",
    "                                    for r12, columnNumber=2\n",
    "                                    for r0, columnNumber=1\n",
    "    \n",
    "    Return:\n",
    "        array (list of list) : sorted list of triplets and their positions\n",
    "\n",
    "    \"\"\"\n",
    "    if len(array) == 0:\n",
    "        return array\n",
    "    # Perform counting sort on each column, starting at the last\n",
    "\n",
    "    column = column_number\n",
    "    while column>=0: \n",
    "        array = counting_sort_by_digit_with_p12(array, alphabet, column)\n",
    "        column-=1 #Switches column in array\n",
    "\n",
    "    return array\n",
    "\n",
    "def counting_sort_by_digit_with_p12(array, alphabet, column):\n",
    "    \"\"\"\n",
    "    Sorts parameter \"array\" using Counting Sort\n",
    "    \n",
    "    Args:\n",
    "        array (list of list): list of triplets and their positions.\n",
    "        \n",
    "        alphabet (dictionnary): alphabet of our array.O uput of function \"alphabetT\" or \"alphabetR0_with_p12\"\n",
    "        \n",
    "        columnNumber (int): number of last column of the lists in our paramater \"array\". \n",
    "                                    for r12, columnNumber=2\n",
    "                                    for r0, columnNumber=1\n",
    "    \n",
    "    Return:\n",
    "        array (list of list) : sorted list of triplets and their positions according to column number \"column\"\n",
    "\n",
    "    \"\"\"\n",
    "    count_index = -1\n",
    "    count = [0] * len(alphabet)\n",
    "    output = [None] * len(array)\n",
    "\n",
    "  # Count frequencies\n",
    "    for i in range(0, len(array)):\n",
    "        count_index = alphabet[array[i][0][column]]\n",
    "        count[count_index] += 1\n",
    "\n",
    "  # Compute cumulates\n",
    "    for i in range(1, len(alphabet)):\n",
    "        count[i] += count[i - 1]\n",
    "\n",
    "  # Move records\n",
    "    for i in range(len(array) - 1, -1, -1):\n",
    "        count_index = alphabet[array[i][0][column]]\n",
    "        count[count_index] -= 1\n",
    "        output[count[count_index]] = array[i]\n",
    "       \n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1339537d",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "\"sort\" et \"CountingSortByDigits\" réalisent ensemble un tri du type \"radix sort\".\n",
    "\"Radix sort\" a donc une complexité de: $O(d*n)$, avec d=3 le nombre de colonnes et n est la taille de array (ici la taille de r12 ou r0)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b13f245e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def triplets_are_equal(a,b):\n",
    "    \"\"\"\n",
    "    Checks if a and b have the same elements.\n",
    "    \n",
    "    Args:\n",
    "        a (list of int) \n",
    "        \n",
    "        b (list of int)\n",
    "\n",
    "    Return:\n",
    "        (bool): True if the triplets are equal, False if not\n",
    "\n",
    "    \"\"\"\n",
    "    for i in range(len(a)):\n",
    "        if a[i]!=b[i]:\n",
    "            return False\n",
    "    return True\n",
    "\n",
    "\n",
    "def ordre_with_p12(r12_p12_sorted, use_index_12=False):\n",
    "    \"\"\"\n",
    "    Returns the order of each element of parameter \"R12p12sorted\".\n",
    "    If use_index_12 is set to \"True\", it also returns a dictionnary, with the position of the triplet\n",
    "    as a key, and its order as element.\n",
    "    \n",
    "    Example: \n",
    "        ordre_with_p12([ [ [0,0,0],1], [ [0,0,0],4], [ [0,0,1],7])=[1,1,2]\n",
    "        \n",
    "        ordre_with_p12([ [ [0,0,0],1], [ [0,0,0],4], [ [0,0,1],7], use_index_12=True)= [ [1,1,2], {1:1,4:1,7:2} ]\n",
    "    \n",
    "    args : \n",
    "        r12_p12_sorted (list) : list of triplets with their position in the sequence, ordered \n",
    "        \n",
    "        use_index_12 (boolean) : True if you want a dictionnary back\n",
    "        \n",
    "    return : \n",
    "        if use_index_12=True : \n",
    "            order (list of int) :  order of each element of parameter \"r12_p12_sorted\"\n",
    "            \n",
    "            repetition (boolean) : If two triplets have the same ordre, \"repetition\"=True\n",
    "            \n",
    "            indexdict (dictionary) :keys are the positions p12 and elements are the order of the positions \n",
    "        \n",
    "        if use_index_12=False : \n",
    "            order (list of int): order of each element of parameter \"r12_p12_sorted\"\n",
    "    \"\"\"\n",
    "    # nous renovie la liste order du genre (1,2,2,3,4,4,5)  et un booléen indiquant s'il y a répétition\n",
    "    index=1\n",
    "    repetition=False\n",
    "    order=[1]\n",
    "    \n",
    "    if use_index_12:\n",
    "        index_dict={r12_p12_sorted[0][1]:1}\n",
    "        \n",
    "    for i in range(1, len(r12_p12_sorted)): \n",
    "        if triplets_are_equal(r12_p12_sorted[i-1][0], r12_p12_sorted[i][0]):\n",
    "            order.append(index)\n",
    "            repetition=True\n",
    "        else : \n",
    "            index+=1\n",
    "            order.append(index)\n",
    "            \n",
    "        if use_index_12:\n",
    "            index_dict[r12_p12_sorted[i][1]]=order[i]\n",
    "    if use_index_12:\n",
    "        return order, repetition, index_dict\n",
    "    else: \n",
    "        return order\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c90f10e",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "triplets_are_equal a une complexité de $O(1)$ donc ordre_with_p12 a une complexité de $O(n)$ à cause la boucle for, n étant la taille de r12_p12_sorted. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "992507a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def alphabetT(T):\n",
    "    \"\"\"\n",
    "    Returns a dictionnary with the order of each \"letter\" constituting the parameter T.\n",
    "    \n",
    "    Example: \n",
    "        alphabetT([4,9,14,67])={4:0, 9:1, 14:2, 67:3}\n",
    "    \n",
    "    Args:\n",
    "        T (list of int): the sequence we want a suffix array from\n",
    "    \n",
    "    Return:\n",
    "        dic (dictionnary) : order of each \"letter\" constituting the parameter T\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    dic={}\n",
    "    a=[]\n",
    "    for i in range(len(T)):\n",
    "        a.append(T[i])\n",
    "    a.sort() \n",
    "    \n",
    "    for element in a:\n",
    "        if not (element in dic):\n",
    "            dic[element]=len(dic)\n",
    "    return dic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "51c44263",
   "metadata": {},
   "outputs": [],
   "source": [
    "def alphabet_r0_with_p12(r0_p0):\n",
    "    \"\"\"\n",
    "    Returns a dictionnary with the order of each \"letter\" constituting the parameter r0_p0.\n",
    "    Similar to the function alphabetT, coded specifically for the output of function position0_R0_p0\n",
    "    \n",
    "    Example: \n",
    "        alphabet_r0_with_p12([[[65, 14], 0], [[67, 11], 3]]) = {11: 0, 14: 1, 65: 2, 67: 3}\n",
    "    \n",
    "    Args:\n",
    "        r0_p0 (list of list): list of couples and their position in the sequence we want a suffix array from.\n",
    "                             output of function position0_R0_p0\n",
    "                             [ [couple], position multiple of 3]\n",
    "    \n",
    "    Return:\n",
    "        dic (dictionnary) : order of each \"letter\" constituting the parameter r0_p0\n",
    "\n",
    "    \"\"\"\n",
    "    \n",
    "    dic={}\n",
    "    a=[]\n",
    "    for column in range(2):\n",
    "        for i in range(len(r0_p0)):\n",
    "            a.append(r0_p0[i][0][column])\n",
    "    a.sort() \n",
    "    \n",
    "    for element in a:\n",
    "        if not (element in dic):\n",
    "            dic[element]=len(dic)\n",
    "    return dic"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "374adf16",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "alphabetT a une complexité de $O(n)$(n étant la taille de 'T') car la fonction parcourt toute la séquence une fois. Il en est de même pour alphabetR0_with_p12, car la seconde boucle for n'est que de taille 2. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2237070",
   "metadata": {},
   "source": [
    "### Deuxième étape de DC3: utilisation des premières positions triées pour construire second groupe de positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "a616f4b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def Tprime_with_p12(p12, index_dict):\n",
    "    \"\"\"\n",
    "    Return a list of the order of each element of p12.\n",
    "    The list will be the new sequence from which we want to extract a suffix array.\n",
    "    \n",
    "    Args : \n",
    "        p12 (list of int) : positions not multiple of 3\n",
    "        \n",
    "        index_dict (dictionnary) : keys are the positions p12 and elements are the order of the positions \n",
    "        \n",
    "    Return :\n",
    "        t(list of int): order of each element of p12\n",
    "    \"\"\"\n",
    "    t=[]\n",
    "    for p in p12:\n",
    "        t.append(index_dict[p])\n",
    "    return t"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "301ce172",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "Parcourir la liste des positions p12 une fois implique que Tprime_with_p12 a une compléxité $O(n)$, n étant la taille de p12."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "679e86bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def position0_R0_p0(T, index_12_dict):\n",
    "    \"\"\"\n",
    "    Returns:\n",
    "    -list of positions multiples of 3\n",
    "    -list of lists, composed of: 1/ [element at position multiple of 3, order of the element at next position]\n",
    "                                2/ position multiple of 3\n",
    "    Args : \n",
    "        T (list of int) : sequence for which we are looking to extract the suffix array. \n",
    "                        It must contain the three sentinel numbers\n",
    "        \n",
    "        index_12_dict (dictionary) : dictionary which contains the positions of p12 as key and their order as elements\n",
    "        \n",
    "    Return :\n",
    "        position (list): all positions multiple of 3\n",
    "        \n",
    "        R (list of list): lists each position with the elements at the next position\n",
    "\n",
    "    \"\"\"\n",
    "    position=[]\n",
    "    R=[]\n",
    "    for i in range(len(T)-3): \n",
    "        if i%3==0:\n",
    "            position.append(i)\n",
    "            if i+1<len(T)-3:\n",
    "                R.append([[T[i],index_12_dict[i+1]],position[-1]])\n",
    "            else:\n",
    "                R.append([[T[i],1],position[-1]]) \n",
    "    return position, R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "84b828f3",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "Parcourir notre séquence initiale 1 fois implique que position0_R0_p0 a une compléxité $O(n)$, n étant la taille de 'T'-3. La tache \"chercher le rang de la position suivante\" a une complexité de $O(1)$ car on cherche une clé dans un dictionnaire."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c75ddd35",
   "metadata": {},
   "source": [
    "### Troisième étape de DC3: combiner les positions triées pour trouver le suffix array."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "83cfeabf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def merge_with_p12(Tfinal, r0_p0_sorted, index_12_dict) :\n",
    "    \"\"\"\n",
    "    Constructs a list of the positions in the right order to construct the suffix table of parameter \"Tfinale\"\n",
    "    \n",
    "    Args : \n",
    "        Tfinal(list of int) : sequence for which we want to create a suffix table \n",
    "        \n",
    "        r0_p0_sorted(list of list) : sorted output of positionR0P0 \n",
    "        \n",
    "        index_12_dict(dictionary):  dictionary which contains the positions of p12 as key and their order as elements\n",
    "    \n",
    "    Return : \n",
    "        liste (list of int) : positions in the right order to construct the suffix table \n",
    "    \"\"\"\n",
    "\n",
    "    index_12_dict_keys=list(index_12_dict.keys())\n",
    "    liste=[]\n",
    "    A=0\n",
    "    B=0\n",
    "    while A<len(r0_p0_sorted) and B<len(index_12_dict_keys):\n",
    "        a=r0_p0_sorted[A][1]\n",
    "        b=index_12_dict_keys[B]\n",
    "        if Tfinal[a]!=Tfinal[b] :\n",
    "            minimum=min(Tfinal[a], Tfinal[b])\n",
    "            \n",
    "            if minimum == Tfinal[a]:\n",
    "                A+=1\n",
    "                liste.append(a)\n",
    "            else: \n",
    "                B+=1\n",
    "                liste.append(b)\n",
    "\n",
    "        else :\n",
    "            if b%3==1 : \n",
    "                longueur=len(liste)\n",
    "                i=0\n",
    "                if index_12_dict[a+1]<index_12_dict[b+1]:\n",
    "                    liste.append(a)\n",
    "                    A+=1\n",
    "                else:\n",
    "                    liste.append(b)\n",
    "                \n",
    "                    B+=1\n",
    "                    \n",
    "                    \n",
    "            elif b%3==2 :\n",
    "            \n",
    "                if Tfinal[a+1]!=Tfinal[b+1] :\n",
    "                   \n",
    "                    minimum=min(Tfinal[a+1], Tfinal[b+1])\n",
    "                    if minimum == Tfinal[a+1]:\n",
    "                        A+=1\n",
    "                    \n",
    "                        liste.append(a)\n",
    "                    else: \n",
    "                        B+=1\n",
    "                        \n",
    "                        liste.append(b)\n",
    "\n",
    "                else:\n",
    "                  \n",
    "                    if index_12_dict[a+2]<index_12_dict[b+2]:\n",
    "                        liste.append(a)\n",
    "                        A+=1\n",
    "                    else:\n",
    "                        liste.append(b)\n",
    "                        B+=1\n",
    "                        \n",
    "\n",
    "    if A==len(r0_p0_sorted):\n",
    "\n",
    "        for i in range(B,len(index_12_dict_keys)):\n",
    "            liste.append(index_12_dict_keys[i])\n",
    "                \n",
    "    if B==len(index_12_dict_keys):\n",
    "\n",
    "        for i in range(A, len(r0_p0_sorted)):\n",
    "            liste.append(r0_p0_sorted[i][1])\n",
    "\n",
    "    return liste"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "235324b1",
   "metadata": {},
   "source": [
    "##### Complexité:\n",
    "Notre fonction est constitué d'une \"boucle while\", puis de deux \"boucles for\". Mais une seule des deux \"boucles for\" sera parcourue, car elles sont précédées par des tests conditionnels: si un \"if\" est entré, le second ne pourra pas l'être.  \n",
    "La \"boucle while\" suivit de n'importe laquelle des deux \"boucles for\" ne parcourt que n éléments, n étant la taille de 'r0_p0_sorted'. Les boucles ne sont pas imbriquées les unes dans les autres, et dans la \"boucle while\", seules des opérations de complexité $O(1)$ prennent place.\n",
    "Donc merge_with_p12 a une complexité de $O(n)$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5357493c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def remove_sentinel(index):\n",
    "    \"\"\"\n",
    "    Return our suffix array without the position of the three sentinels numbers\n",
    "    \n",
    "    Args : \n",
    "        index (list of int) : list of indexes \n",
    "        \n",
    "    Return :\n",
    "        index[1:](list of int) : list of indexes without the sentinels\n",
    "    \n",
    "    \"\"\"\n",
    "    return index[1:]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d679e6c1",
   "metadata": {},
   "source": [
    "##### Complexité\n",
    "removesentinel a une complexité temporelle de $0(1)$."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "cf1426fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def resumeHigherOrder_with_p12(index_012_prime, P12):\n",
    "    \"\"\"\n",
    "    Returns a dictionnary with positions not multiple of 3 \"P12\" as key, and their order with \n",
    "    the positions of the suffix array \"index012prime\" as elements. It is used to create a new \"index_12_dict\" \n",
    "    element\n",
    "    \n",
    "    Args : \n",
    "        index012prime (list of int) : suffix array of the positions not multiple of 3\n",
    "        \n",
    "        P12 (list of int) : positions not multiple of 3\n",
    "    \n",
    "    Return : \n",
    "        output (dictionnary) : positions not multiple of 3 \"P12\" as key, and their order with the positions of the suffix array \"index012prime\" as elements.\n",
    "    \"\"\"\n",
    "    output={}\n",
    "    for i in range(len(index_012_prime)):\n",
    "        output[P12[index_012_prime[i]]]=i\n",
    "        \n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcc53ec5",
   "metadata": {},
   "source": [
    "##### Complexité\n",
    "resumeHigherOrder_with_p12 parcourt chaque élément de index_012_prime, qui est constitué de 2/3 des éléments de notre séquence d'ADN initiale de taille n. \n",
    "La fonction est donc de complexité $0(n)$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "337a7274",
   "metadata": {},
   "source": [
    "### Dernière étape de DC3: assemblage final de la récursion."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "c6a79b12",
   "metadata": {},
   "outputs": [],
   "source": [
    "def almost_dc3_with_p12(T):\n",
    "    \"\"\"\n",
    "    Return the suffix array of the sequence T\n",
    "    \n",
    "    Args : \n",
    "        T (list of int) : list of int corresponding to the ascii code of the initial sequence\n",
    "        \n",
    "    Return : \n",
    "        index_012 (list of int) : positions in order to construct the suffix array of T\n",
    "    \"\"\"\n",
    "    \n",
    "    column_number=2\n",
    "    p12=position1_2(T)\n",
    "    r12=radix_with_p12(p12,T)\n",
    "    alphabet_T=alphabetT(T)\n",
    "    r12_sorted=sort_with_p12(r12, alphabet_T, column_number)    \n",
    "    order12,repetition, index_12_dict=ordre_with_p12(r12_sorted, True)\n",
    "    if repetition:\n",
    "        Tprim=Tprime_with_p12(p12, index_12_dict)+[0,0,0]\n",
    "        index_012=almost_dc3_with_p12(Tprim)\n",
    "        index_12_dict=resumeHigherOrder_with_p12(index_012, p12)\n",
    "    p0,r0_p0=position0_R0_p0(T, index_12_dict)\n",
    "    alphabet_r0=alphabet_r0_with_p12(r0_p0)\n",
    "    r0_sorted=sort_with_p12(r0_p0,alphabet_r0 ,column_number-1)\n",
    "    index_012=remove_sentinel(merge_with_p12(T, r0_sorted, index_12_dict))\n",
    "    \n",
    "    return index_012\n",
    "   \n",
    "    \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "700190ab",
   "metadata": {},
   "source": [
    "##### Complexité\n",
    "almost_dc3_with_p12 est une fonction récursive dont chaque sous fonction a une complexité au pire linéaire.\n",
    "Elle est donc de complexité $0(n)$, n étant la taille de la séquence 'T'."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a1b29c2",
   "metadata": {},
   "source": [
    "#### Testons le code avec l'exemple du cours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "56a19fcd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[8, 0, 3, 6, 9, 1, 4, 7, 2, 5]\n",
      "La fonction pour trouver le suffix array de S a pris 0.0001685619354248047 secondes.\n"
     ]
    }
   ],
   "source": [
    "S=\"abcabcacab\"\n",
    "start=time.time()\n",
    "T=almost_dc3_with_p12(asciiDC3(S))\n",
    "end=time.time()\n",
    "column_number=2\n",
    "print(T)\n",
    "print(\"La fonction pour trouver le suffix array de S a pris \"+str(end-start)+\" secondes.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4eaf1a2c",
   "metadata": {},
   "source": [
    "#### Testons le code avec le génome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d589cce0",
   "metadata": {},
   "source": [
    "Suffix table du premier chromosome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "e2b39d26",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pour le premier chromosome, il nous faut 17.118040084838867 secondes pour calculer sa suffix table.\n",
      "Le premier chromosome a une taille de 640851 nucléotides.\n"
     ]
    }
   ],
   "source": [
    "start=time.time()\n",
    "almost_dc3_with_p12(asciiDC3(genome[0].seq.upper()))\n",
    "end=time.time()\n",
    "print(\"Pour le premier chromosome, il nous faut \"+str(end-start)+\" secondes pour calculer sa suffix table.\")\n",
    "print(\"Le premier chromosome a une taille de \"+str(len(genome[0].seq.upper()))+\" nucléotides.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "3f7b3440",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Il nous faut 211.94798851013184 secondes pour calculer la suffix table de 5 chromosome.\n",
      "En moyenne, il nous faut donc 30.278284072875977 secondes pour calculer la suffix table d'un chromosome.\n",
      "Nos 7 chromosomes ont une taille de 5199971 nucléotides.\n"
     ]
    }
   ],
   "source": [
    "start0=time.time()\n",
    "suffix_array_0=almost_dc3_with_p12(asciiDC3(genome[0].seq.upper()))\n",
    "t0=time.time()\n",
    "suffix_array_1=almost_dc3_with_p12(asciiDC3(genome[1].seq.upper()))\n",
    "t1=time.time()\n",
    "suffix_array_2=almost_dc3_with_p12(asciiDC3(genome[2].seq.upper()))\n",
    "t2=time.time()\n",
    "suffix_array_3=almost_dc3_with_p12(asciiDC3(genome[3].seq.upper()))\n",
    "t3=time.time()\n",
    "suffix_array_4=almost_dc3_with_p12(asciiDC3(genome[4].seq.upper()))\n",
    "t4=time.time()\n",
    "suffix_array_5=almost_dc3_with_p12(asciiDC3(genome[5].seq.upper()))\n",
    "t5=time.time()\n",
    "suffix_array_6=almost_dc3_with_p12(asciiDC3(genome[6].seq.upper()))\n",
    "end=time.time()\n",
    "print(\"Il nous faut \"+str(end-start)+\" secondes pour calculer la suffix table de 5 chromosome.\")\n",
    "print(\"En moyenne, il nous faut donc \"+str((end-start)/7)+\" secondes pour calculer la suffix table d'un chromosome.\")\n",
    "print(\"Nos 7 chromosomes ont une taille de \"+str(len(genome[0].seq.upper())+len(genome[1].seq.upper())+len(genome[2].seq.upper())+len(genome[3].seq.upper())+len(genome[4].seq.upper()))+\" nucléotides.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a28802a2",
   "metadata": {},
   "source": [
    "Visualisons la complexité de la fonction codant pour dc3:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "3d4b8319",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXAAAAEJCAYAAACNNHw2AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAARi0lEQVR4nO3df4zkd33f8efrbNNkC4mP3kKvmL1NqcMPWcmRbCxUqtSYoDpuU+M2aYO21EppN/1BBFVIoTmpEEUn0TaEKEpDtQQrRp2CaCHBpaSpZbAcEn6tk/NhY6emwXs1Pfku/CihW1Fsv/vH97v1en17M3O3szOf2+dDGn3n+/l+Z+bl3fNrvvv9fuc7qSokSe05MO0AkqQLY4FLUqMscElqlAUuSY2ywCWpURa4JDVq5AJPclmSP0jykX7+2UnuSPJQPz04uZiSpO3G2QJ/A/DAlvm3AHdW1dXAnf28JGmPjFTgSa4C/irwa1uGbwJu6+/fBrx6V5NJks7r8hHX+yXgnwHP2jL23Ko6DVBVp5M8Z9iTHDp0qBYXF8fNKEn72j333PPHVTW/fXxogSf5a8CZqronyXXjvnCSFWAFYGFhgbW1tXGfQpL2tSTr5xofZRfKy4G/nuRh4P3A9Un+HfBoksP9kx8GzpzrwVW1WlVLVbU0P/+0NxBJ0gUaWuBV9c+r6qqqWgR+HPhYVf0d4Hbgln61W4APTyylJOlpLuY88LcDr0ryEPCqfl6StEdGPYgJQFXdBdzV3/8y8MrdjyRJGoWfxJSkRlngkjRJgwEsLsKBA910MNi1px5rF4okaQyDAayswMZGN7++3s0DLC9f9NO7BS5Jk3Ls2JPlvWljoxvfBRa4JE3KqVPjjY/JApekSVlYGG98TBa4JE3K8eMwN/fUsbm5bnwXWOCSNCnLy7C6CkeOQNJNV1d35QAmeBaKJE3W8vKuFfZ2boFLUqMscElqlAUuSY2ywCWpURa4JDXKApekRlngktQoC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ4JLUKAtckhplgUtSoyxwSWqUBS5JjbLAJalRFrgkNcoCl6RGDS3wJN+W5DNJ7k1yf5Kf68ffluRLSU70txsnH1eStOnyEdb5JnB9VX0jyRXAJ5L8Vr/snVX1C5OLJ0naydACr6oCvtHPXtHfapKhJEnDjbQPPMllSU4AZ4A7qurT/aLXJzmZ5NYkBycVUpL0dCMVeFU9XlVHgauAa5NcA7wLeAFwFDgNvONcj02ykmQtydrZs2d3JbQkacyzUKrqa8BdwA1V9Whf7E8A7wau3eExq1W1VFVL8/PzF5tXkqZvMIDFRThwoJsOBlOJMcpZKPNJruzvfzvwQ8CDSQ5vWe1m4L6JJJSkWTIYwMoKrK9DVTddWZlKiY+yBX4Y+HiSk8Bn6faBfwT4V0k+14+/AvinE8wpSbPh2DHY2Hjq2MZGN77HRjkL5STw0nOMv3YiiSRplp06Nd74BPlJTEkax8LCeOMTZIFL0jiOH4e5uaeOzc1143vMApekcSwvw+oqHDkCSTddXe3G99goH6WXJG21vDyVwt7OLXBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ4JLUKAtckhplgUtSoyxwSWqUBS5JjbLAJalRFrgkNcoCl6RGWeCS1CgLXJIaZYFLUqMscElqlAUuSY2ywCWpURa4JDXKApekRlng0n4yGMDiIhw40E0Hg2kn0kW4fNoBJO2RwQBWVmBjo5tfX+/mAZaXp5dLF8wtcGm/OHbsyfLetLHRjatJFri0X5w6Nd64Zp4FLu0XCwvjjWvmDS3wJN+W5DNJ7k1yf5Kf68efneSOJA/104OTjyvpgh0/DnNzTx2bm+vGW7dPD86OsgX+TeD6qvpe4ChwQ5KXAW8B7qyqq4E7+3lJs2p5GVZX4cgRSLrp6mr7BzA3D86ur0PVkwdn90GJp6pGXzmZAz4B/CPgvcB1VXU6yWHgrqp64fkev7S0VGtraxeTV5KeanGxK+3tjhyBhx/e6zQTkeSeqlraPj7SPvAklyU5AZwB7qiqTwPPrarTAP30ObuYV5JGs48Pzo5U4FX1eFUdBa4Crk1yzagvkGQlyVqStbNnz15gTEnawT4+ODvWWShV9TXgLuAG4NF+1wn99MwOj1mtqqWqWpqfn7+4tJK03aV8cHaIUc5CmU9yZX//24EfAh4Ebgdu6Ve7BfjwhDJK0s4u1YOzIxjlo/SHgduSXEZX+B+oqo8k+STwgSSvA04BPzbBnJK0s+XlfVHY2w0t8Ko6Cbz0HONfBl45iVCSpOH8JKYkNcoCl6RGWeCS1CgLXJIaZYFLUqMscElqlAUuSY2ywCWpURa4JDXKApekRlngktQoC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ4NIoBgNYXIQDB7rpYDDtRNJIX2os7W+DAayswMZGN7++3s3DvvwiXc0Ot8ClYY4de7K8N21sdOPSFFng0jCnTo03Lu0RC1waZmFhvHFpj1jg0jDHj8Pc3FPH5ua6cWmKLHBpmOVlWF2FI0cg6aarqx7A1NR5Foo0iuVlC1szxy1wSWqUBS5JjbLAJalRFrgkNcoCl6RGDS3wJM9P8vEkDyS5P8kb+vG3JflSkhP97cbJx5UkbRrlNMLHgJ+uqt9P8izgniR39MveWVW/MLl4kqSdDC3wqjoNnO7v/0mSB4DnTTqYJOn8xtoHnmQReCnw6X7o9UlOJrk1ycHdDidJ2tnIBZ7kmcAHgTdW1deBdwEvAI7SbaG/Y4fHrSRZS7J29uzZi08sSQJGLPAkV9CV96CqPgRQVY9W1eNV9QTwbuDacz22qlaraqmqlubn53crtyTte6OchRLgPcADVfWLW8YPb1ntZuC+3Y8nSdrJKGehvBx4LfC5JCf6sZ8FXpPkKFDAw8BPTiCfJGkHo5yF8gkg51j00d2PI0kalZ/ElKRGWeCS1CgLXJIaZYFLUqMscElqlAUuSY2ywCWpURa4JDXKApekRlngktQoC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ4JLUKAtckhplgUtSoyxwSWqUBS5JjbLAJalRFrgkNcoCl6RGWeCS1CgLXJIaZYFLUqMscElqlAUuSY2ywCWpUUMLPMnzk3w8yQNJ7k/yhn782UnuSPJQPz04+biSpE2jbIE/Bvx0Vb0YeBnwT5K8BHgLcGdVXQ3c2c9L5zcYwOIiHDjQTQeDaSeSmjW0wKvqdFX9fn//T4AHgOcBNwG39avdBrx6Qhl1qRgMYGUF1tehqpuurFji0gVKVY2+crII3A1cA5yqqiu3LPtqVT1tN0qSFWAFYGFh4fvX19cvMrKatbjYlfZ2R47Aww/vdRqpGUnuqaql7eMjH8RM8kzgg8Abq+rroz6uqlaraqmqlubn50d9mC5Fp06NNy7pvEYq8CRX0JX3oKo+1A8/muRwv/wwcGYyEXXJWFgYb1zSeY1yFkqA9wAPVNUvbll0O3BLf/8W4MO7H0+XlOPHYW7uqWNzc924pLGNsgX+cuC1wPVJTvS3G4G3A69K8hDwqn5e2tnyMqyudvu8k266utqNSxrbWAcxL9bS0lKtra3t2etJ0qXgog9iSpJmiwUuSY2ywCWpURa4JDXKApekRlngktQoC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEZZ4JLUKAtckhplgUtSoyxwSWqUBS5JjbLAJalRFrgkNcoCl6RGWeCS1CgLXJIaZYFLUqMscElqlAUuSY2ywCWpURa4JDXKApekRlngktSooQWe5NYkZ5Lct2XsbUm+lOREf7txsjElSduNsgX+68AN5xh/Z1Ud7W8f3d1YkqRhhhZ4Vd0NfGUPskiSxnAx+8Bfn+Rkv4vl4K4lkiSN5EIL/F3AC4CjwGngHTutmGQlyVqStbNnz17gy0mStrugAq+qR6vq8ap6Ang3cO151l2tqqWqWpqfn7/QnJKkbS6owJMc3jJ7M3DfTutKkibj8mErJHkfcB1wKMkjwFuB65IcBQp4GPjJyUWUJJ3L0AKvqtecY/g9E8giSRqDn8SUpEbNfoEPBrC4CAcOdNPBYNqJJGkmDN2FMlWDAayswMZGN7++3s0DLC9PL5ckzYDZ3gI/duzJ8t60sdGNS9I+N9sFfurUeOOStI/MdoEvLIw3Lkn7yGwX+PHjMDf31LG5uW5ckva52S7w5WVYXYUjRyDppqurHsCUJGb9LBToytrClqSnme0tcEnSjixwSWqUBS5JjbLAJalRFrgkNSpVtXcvlpwF1vfsBUd3CPjjaYfYZhYzgbnGZa7xzGKuWch0pKqe9pVme1rgsyrJWlUtTTvHVrOYCcw1LnONZxZzzWKmTe5CkaRGWeCS1CgLvLM67QDnMIuZwFzjMtd4ZjHXLGYC3AcuSc1yC1ySGrVvCjzJDUn+MMkXkrzlHMt/JsmJ/nZfkseTPHsGcn1nkv+U5N4k9yf5iUlnGjHXwSS/keRkks8kuWYPMt2a5EyS+3ZYniS/3Gc+meT7Jp1pxFwvSvLJJN9M8qa9yDRiruX+53Qyye8l+d4ZyXVTn+lEkrUkf2kWcm1Z7wf6fvjRvch1XlV1yd+Ay4D/Dvx54BnAvcBLzrP+jwAfm4VcwM8C/7K/Pw98BXjGDOT618Bb+/svAu7cg5/XDwLfB9y3w/Ibgd8CArwM+PQe/fsalus5wA8Ax4E37UWmEXP9ReBgf/+HZ+jn9Uye3L37PcCDs5CrX+cy4GPAR4Ef3avf5U63/bIFfi3whar6o6r6v8D7gZvOs/5rgPfNSK4CnpUkdP+wvwI8NgO5XgLcCVBVDwKLSZ47yVBVdTfdf/9ObgLeW51PAVcmOTzJTKPkqqozVfVZ4FuTzrLtdYfl+r2q+mo/+yngqhnJ9Y3q2xL403T/D0w9V++ngA8CZyafaLj9UuDPA/7HlvlH+rGnSTIH3ED3S5qFXL8CvBj4n8DngDdU1RMzkOte4G8AJLkWOMIeFcB5jPx71tO8ju6vl5mQ5OYkDwL/Gfh7084DkOR5wM3Av512lk37pcBzjrGd3tV/BPjdqhr2TrwbRsn1V4ATwJ8DjgK/kuQ7JhtrpFxvBw4mOUG3VfIHTP4vg2HG+T2rl+QVdAX+5mln2VRVv1FVLwJeDfz8lONs+iXgzVX1+LSDbJr9b+TZHY8Az98yfxXdFu25/Dh7s/sERsv1E8Db+z8pv5Dki3T7nD8zzVxV9fU+G/3unS/2t2ka5/csIMn3AL8G/HBVfXnaebarqruTvCDJoaqa9vVIloD3d//cOQTcmOSxqvrNaQXaL1vgnwWuTvJdSZ5BV9K3b18pyXcCfxn48AzlOgW8ss/3XOCFwB9NO1eSK/tlAH8fuLsv9Wm6Hfi7/dkoLwP+V1WdnnKmmZVkAfgQ8Nqq+m/TzrMpyV/oNwrozyR6BjD1N5eq+q6qWqyqReA/Av94muUN+2QLvKoeS/J64LfpjiLfWlX3J/mH/fLNfVo3A/+1qv73DOX6eeDXk3yObhfBmye9JTJirhcD703yOPB5uj/BJyrJ+4DrgENJHgHeClyxJdNH6c5E+QKwQf8XwrRzJfmzwBrwHcATSd5Id1bPRN/wRvh5/QvgzwC/2vflY7UHF20aIdffpHsj/hbwf4C/veWg5jRzzRw/iSlJjdovu1Ak6ZJjgUtSoyxwSWqUBS5JjbLAJWlCRr1A1pb1/1aSz/cXrvv3Q9f3LBRJmowkPwh8g+4aPee9YmeSq4EPANdX1VeTPKeqznvNFbfAJWlCznWBrP6Tpf8lyT1JfifJi/pF/wD4N5sXGBtW3mCBS9JeWwV+qqq+H3gT8Kv9+HcD353kd5N8KskNw55oX3wSU5JmQZJn0l2H/T/0n34F+FP99HLgarpPg14F/E6Sa6rqazs9nwUuSXvnAPC1qjp6jmWPAJ+qqm8BX0zyh3SF/tnzPZkkaQ/017/5YpIfg///NYCbX2X3m8Ar+vFDdLtUznvhOgtckiakv0DWJ4EXJnkkyeuAZeB1Se4F7ufJb7v6beDLST4PfBz4mWGX+PU0QklqlFvgktQoC1ySGmWBS1KjLHBJapQFLkmNssAlqVEWuCQ1ygKXpEb9PwDAjYMgXszbAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.plot([len(genome[0].seq.upper()), len(genome[1].seq.upper()), len(genome[2].seq.upper()),len(genome[3].seq.upper()), len(genome[4].seq.upper()), len(genome[5].seq.upper()), len(genome[6].seq.upper())], [t0-start0,t1-t0, t2-t1, t3-t2,t4-t3,t5-t4, end-t5], 'ro')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e839251",
   "metadata": {},
   "source": [
    "On observe des linéarités discontinues: la complexité semble linéaire entre le 2eme et 4eme chromosome, ainsi qu'entre le 5eme et le 7eme."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bb10b263",
   "metadata": {},
   "source": [
    "Mesurons le temps que prend chaque sous fonction de notre algorithme DC3 grâce à la bibliothèque cProfile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "a111f300",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fri Nov 10 21:02:30 2023    dc3_stats\n",
      "\n",
      "         21257127 function calls (21257121 primitive calls) in 17.860 seconds\n",
      "\n",
      "   Ordered by: cumulative time\n",
      "\n",
      "   ncalls  tottime  percall  cumtime  percall filename:lineno(function)\n",
      "        1    0.000    0.000   17.860   17.860 {built-in method builtins.exec}\n",
      "        1    0.279    0.279   17.859   17.859 <string>:1(<module>)\n",
      "      7/1    0.681    0.097   17.391   17.391 /tmp/ipykernel_36096/3977718216.py:1(almost_dc3_with_p12)\n",
      "       14    0.096    0.007    5.279    0.377 /tmp/ipykernel_36096/2430097404.py:1(sort_with_p12)\n",
      "       35    5.182    0.148    5.182    0.148 /tmp/ipykernel_36096/2430097404.py:29(counting_sort_by_digit_with_p12)\n",
      "        7    3.238    0.463    3.330    0.476 /tmp/ipykernel_36096/3017114688.py:1(position0_R0_p0)\n",
      "        7    2.489    0.356    2.912    0.416 /tmp/ipykernel_36096/3105641319.py:1(merge_with_p12)\n",
      "        7    1.017    0.145    1.746    0.249 /tmp/ipykernel_36096/2535467419.py:20(ordre_with_p12)\n",
      "        7    0.786    0.112    0.848    0.121 /tmp/ipykernel_36096/762813728.py:1(radix_with_p12)\n",
      "        7    0.475    0.068    0.774    0.111 /tmp/ipykernel_36096/2356060223.py:1(alphabet_r0_with_p12)\n",
      "        7    0.403    0.058    0.704    0.101 /tmp/ipykernel_36096/3536684344.py:1(alphabetT)\n",
      "  1206689    0.603    0.000    0.664    0.000 /tmp/ipykernel_36096/2535467419.py:1(triplets_are_equal)\n",
      " 11463619    0.575    0.000    0.575    0.000 {method 'append' of 'list' objects}\n",
      "  7301050    0.407    0.000    0.407    0.000 {built-in method builtins.len}\n",
      "        7    0.336    0.048    0.394    0.056 /tmp/ipykernel_36096/3476465919.py:1(position1_2)\n",
      "        6    0.382    0.064    0.382    0.064 /tmp/ipykernel_36096/2870284556.py:1(resumeHigherOrder_with_p12)\n",
      "       14    0.370    0.026    0.370    0.026 {method 'sort' of 'list' objects}\n",
      "        6    0.265    0.044    0.320    0.053 /tmp/ipykernel_36096/3446150077.py:1(Tprime_with_p12)\n",
      "        1    0.129    0.129    0.188    0.188 /tmp/ipykernel_36096/4151255859.py:1(asciiDC3)\n",
      "   644760    0.094    0.000    0.094    0.000 {built-in method builtins.min}\n",
      "   640851    0.031    0.000    0.031    0.000 {built-in method builtins.ord}\n",
      "        7    0.022    0.003    0.022    0.003 /tmp/ipykernel_36096/3663174410.py:1(remove_sentinel)\n",
      "        1    0.000    0.000    0.000    0.000 /home/azarkua/.local/lib/python3.10/site-packages/Bio/Seq.py:1233(upper)\n",
      "        1    0.000    0.000    0.000    0.000 {method 'upper' of 'bytes' objects}\n",
      "        1    0.000    0.000    0.000    0.000 /home/azarkua/.local/lib/python3.10/site-packages/Bio/Seq.py:466(__iter__)\n",
      "        1    0.000    0.000    0.000    0.000 {method 'decode' of 'bytes' objects}\n",
      "        7    0.000    0.000    0.000    0.000 {method 'keys' of 'dict' objects}\n",
      "        1    0.000    0.000    0.000    0.000 /home/azarkua/.local/lib/python3.10/site-packages/Bio/Seq.py:1961(__init__)\n",
      "        2    0.000    0.000    0.000    0.000 /usr/lib/python3/dist-packages/cycler.py:239(__iter__)\n",
      "        1    0.000    0.000    0.000    0.000 {method 'disable' of '_lsprof.Profiler' objects}\n",
      "        1    0.000    0.000    0.000    0.000 /home/azarkua/.local/lib/python3.10/site-packages/Bio/SeqRecord.py:334(<lambda>)\n",
      "        1    0.000    0.000    0.000    0.000 {built-in method builtins.isinstance}\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<pstats.Stats at 0x7ffb652c7910>"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cProfile.run(\"almost_dc3_with_p12(asciiDC3(genome[0].seq.upper()))\", \"dc3_stats\")\n",
    "p = pstats.Stats(\"dc3_stats\")\n",
    "p.sort_stats(\"cumulative\").print_stats()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "316d6bac",
   "metadata": {},
   "source": [
    "On constate que les fonctions les plus consommatrices en temps, sont les fonctions de tri, et le fonctions initiales de séparation des positions."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fbc43e49",
   "metadata": {},
   "source": [
    "## 3/ Transformée de Burrow-Wheeler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "ee199ee7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def BWT_suffix_table(T,end_of_string=False):\n",
    "    \"\"\"\n",
    "    Compute the BWT by calcultating the suffix_table of parameter 'T'\n",
    "    \n",
    "    Args:\n",
    "        T (str): string\n",
    "        end_of_string (char): end of string character to append\n",
    "    \n",
    "    Return:\n",
    "        bwt (str): BWT\n",
    "    \"\"\"\n",
    "    if end_of_string==False:\n",
    "        T += '!'\n",
    "    suffix_array=almost_dc3_with_p12(asciiDC3(T)) \n",
    "    bwt = \"\"\n",
    "    for i in suffix_array:\n",
    "        bwt += T[i-1]\n",
    "    return(bwt)\n",
    "\n",
    "def BWT(T, suffix_table,end_of_string=False):\n",
    "    \"\"\"\n",
    "    Compute the BWT without calculating the suffix table\n",
    "    \n",
    "    Args:\n",
    "        T (str): string\n",
    "        end_of_string (char): end of string character to append\n",
    "        suffix_table (list of int): suffix table of parameter 'T'\n",
    "    \n",
    "    Return:\n",
    "        bwt (str): BWT\n",
    "    \"\"\"\n",
    "    if end_of_string==False:\n",
    "        T += '!'\n",
    "    bwt = \"\"\n",
    "    for i in suffix_table:\n",
    "        bwt += T[i-1]\n",
    "    return(bwt)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b73c1dc5",
   "metadata": {},
   "source": [
    "# ATTENTION !!!! dans BWT, tu dois enlever end_of_string et toujours mettre des séquences avec un ! à la fin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "020d2a41",
   "metadata": {},
   "source": [
    "##### Complexité : \n",
    "BWT_suffix_table utilise almost_dc3_with_p12 (qui a une complexité de $0(n)$, n étant la taille de 'T') et une boucle for qui parcourt suffix array (ce qui donne une compléxité de $0(m)$, m étant la taille de 'suffix_array'). Elle ajoute un élément à la fin de notre séquence d'ADN, qui sera donc de taille n+1.\n",
    "Cela revient donc à une complexité de $0(n+1+m)$. Comme 'suffix_array' est la table de suffixe de T, elle est donc aussi de taille n. On a donc une complexité de $0(2n+1))$\n",
    "\n",
    "BWT parcourt seulement 'suffix_table', elle a donc une complexité de $0(n)$, avec n la taille de la séquence T donnée en paramètre."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99481c90",
   "metadata": {},
   "source": [
    "### Cherchons la transformée de Burrow-Wheeler d'un string simple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "50ee964b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "test_2='ATGCTAGCTGCCCTGATCTCTCTGA!'\n",
    "suffix_array_2_with_p12=almost_dc3_with_p12(asciiDC3(test_2)) "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c865d7e",
   "metadata": {},
   "source": [
    "Visualisons la suffix table ainsi que la BWT de la séquence \"test_2\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "2d06b591",
   "metadata": {},
   "outputs": [],
   "source": [
    "def print_suffix_table(sequence, sf, visualize_bwt=False):\n",
    "    '''\n",
    "    Allows visualisation of the suffix_table (parameter sf) of the given sequence (parameter sequence).\n",
    "    If visualize_bwt is set to True, every line of the suffix_table starts with the corresponding\n",
    "    letter of the BWT of parameter \"sequence\".\n",
    "    \n",
    "    Args:\n",
    "        sequence(list of string) : Letter sequence from which we calculated the bwt and suffix table\n",
    "        sf (list of int): positions in \"sequence\" of the first letter of every line of the suffix table\n",
    "        visualize_bwt (bool): allows vizualisation of the BWT if set to \"True\"\n",
    "    \n",
    "    Return:\n",
    "        None\n",
    "    '''\n",
    "    for i in range(len(sf)):\n",
    "        if visualize_bwt:\n",
    "            print(str(sequence[sf[i]-1])+\"-\"+str(sequence[sf[i]:]))\n",
    "        else:\n",
    "            print(str(sequence[sf[i]:]))\n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a191aea",
   "metadata": {},
   "source": [
    "##### Complexité : \n",
    "print_suffix_table parcourt 'sf', elle a donc une complexité de $0(n)$, n étant la taille de la sequence donnée en paramètre. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "dffe1a95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A-!\n",
      "G-A!\n",
      "T-AGCTGCCCTGATCTCTCTGA!\n",
      "G-ATCTCTCTGA!\n",
      "!-ATGCTAGCTGCCCTGATCTCTCTGA!\n",
      "G-CCCTGATCTCTCTGA!\n",
      "C-CCTGATCTCTCTGA!\n",
      "G-CTAGCTGCCCTGATCTCTCTGA!\n",
      "T-CTCTCTGA!\n",
      "T-CTCTGA!\n",
      "T-CTGA!\n",
      "C-CTGATCTCTCTGA!\n",
      "G-CTGCCCTGATCTCTCTGA!\n",
      "T-GA!\n",
      "T-GATCTCTCTGA!\n",
      "T-GCCCTGATCTCTCTGA!\n",
      "T-GCTAGCTGCCCTGATCTCTCTGA!\n",
      "A-GCTGCCCTGATCTCTCTGA!\n",
      "C-TAGCTGCCCTGATCTCTCTGA!\n",
      "A-TCTCTCTGA!\n",
      "C-TCTCTGA!\n",
      "C-TCTGA!\n",
      "C-TGA!\n",
      "C-TGATCTCTCTGA!\n",
      "C-TGCCCTGATCTCTCTGA!\n",
      "A-TGCTAGCTGCCCTGATCTCTCTGA!\n"
     ]
    }
   ],
   "source": [
    "# On visualise à la fois la suffix table, et en première position à gauche, la BWT\n",
    "print_suffix_table(test_2,suffix_array_2_with_p12, visualize_bwt=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d951740a",
   "metadata": {},
   "source": [
    "### Appliquons la transformée de Burrow-Wheeler à notre génome de référence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "e856d22b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Calculer la transformée de Burrow-Wheeler de notre chromosome 1 prend 0.7128598690032959 secondes.\n"
     ]
    }
   ],
   "source": [
    "# Comme nous avons calculé les tables de suffixe précédemment, nous les incluons comme paramètre dans \n",
    "# notre calcul de BWT\n",
    "start=time.time()\n",
    "bwt_T=BWT(genome[0].seq.upper(),suffix_array_0)\n",
    "end=time.time()\n",
    "print(\"Calculer la transformée de Burrow-Wheeler de notre chromosome 1 prend \"+str(end-start)+\" secondes.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6bb32ed0",
   "metadata": {},
   "source": [
    "Mesurons le temps que chaque sous fonction de la fonction \"BWT_suffix_table\" prend:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c34d8171",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAEFCAYAAADzHRw3AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAR9klEQVR4nO3df4xlZ13H8fenW2pcEVrcAbTb3a26UFYCDQ4VjWKVIFsiqShq68YmFdxUKcIfECpN5A+yCQZN/EFxMyENIVnbqPwqsVAT/FGCrexUtj+2ULK22XYspoNVCKyhbPv1j3uH3k5n5p7p3jv37jPvVzKZOec8e++nM81nzpx7z/OkqpAknf7OmHQASdJoWOiS1AgLXZIaYaFLUiMsdElqxJmTeuJt27bVrl27JvX0knRauuOOO75eVTMrHZtYoe/atYv5+flJPb0knZaSHF/tmJdcJKkRFrokNWJooSe5PskjSe5Z5fgFSW5L8p0k7xx9RElSF13O0D8C7F3j+KPAHwB/MopAkqRnZmihV9Wt9Ep7teOPVNVh4LujDCZJWp8NvYaeZH+S+STzi4uLG/nUkjR5hw7Brl1wxhm9z4cOjfThN7TQq2quqmaranZmZsW3UUpSmw4dgv374fhxqOp93r9/pKXuu1wkaSNcey2cOPHUfSdO9PaPiIUuSRvhwQfXt/8ZGHqnaJIbgIuBbUkWgPcCzwKoqoNJXgjMA88BnkjyDmBPVX1zZCkl6XS3Y0fvMstK+0dkaKFX1eVDjv8XsH1kiSSpRQcO9K6ZD1522bq1t39EvOQiSRth3z6Ym4OdOyHpfZ6b6+0fkYlNziVJm86+fSMt8OU8Q5ekRljoktQIC12SGmGhS1IjLHRJaoSFLkmNsNAlqREWuiQ1wkKXpFM15nnOu/JOUUk6FUvznC/N0bI0zzmM9a7QlXiGLkmnYgPmOe/KQpekU7EB85x3ZaFL0qlYbT7zEc5z3pWFLkmn4sCB3rzmg0Y8z3lXFroknYoNmOe8K9/lIkmnaszznHflGbokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqxNBCT3J9kkeS3LPK8ST5iyTHktyV5BWjjylJGqbLGfpHgL1rHL8E2N3/2A/81anHkiSt19BCr6pbgUfXGHIp8NHquR04O8kPjyqgJKmbUVxDPxd4aGB7ob/vaZLsTzKfZH5xcXEETy1Jy0zJYhOTMIpCzwr7aqWBVTVXVbNVNTszMzOCp5akAUuLTRw/DlVPLjaxSUp9FIW+AJw3sL0deHgEjytJ6zNFi01MwigK/Sbgiv67XV4FfKOqvjaCx5Wk9ZmixSYmYehsi0luAC4GtiVZAN4LPAugqg4CNwOvB44BJ4ArxxVWkta0Y0fvMstK+zeBoYVeVZcPOV7AW0eWSJKeqQMHnrpgM0xssYlJ8E5RSe2YosUmJsEFLiS1ZUoWm5gEz9AlqREWuiQ1wkKXpEZY6JqcTXyLtjQOviiqyVi6RXvp7WVLt2jDpn1BSzpVnqFrMjb5LdrSOFjomoxNfou2NA4WuiZjtVuxN8kt2tI4WOiajAMHerdkD9pEt2hL42ChazI2+S3a0jj4LhdNzia+RVsaB8/QJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljo0mblEoDN6VToSfYmuS/JsSTXrHD8nCSfSHJXki8meenoo0oamaUlAI8fh6onlwC01E9rQws9yRbgOuASYA9weZI9y4a9BzhSVS8DrgD+fNRBJY2QSwA2qcsZ+kXAsaq6v6oeA24ELl02Zg/wOYCq+gqwK8kLRppU0ui4BGCTuhT6ucBDA9sL/X2D7gR+FSDJRcBOYPvyB0qyP8l8kvnFxcVnlljSqXMJwCZ1KfSssK+Wbb8fOCfJEeBtwJeAk0/7R1VzVTVbVbMzMzPrzSppVFwCsEldVixaAM4b2N4OPDw4oKq+CVwJkCTAA/0PSdNoaaWoa6/tXWbZsaNX5q4gdVrrUuiHgd1Jzgf+E7gM+K3BAUnOBk70r7G/Bbi1X/KSppVLADZnaKFX1ckkVwO3AFuA66vqaJKr+scPAi8BPprkceBe4M1jzCxJWkGnRaKr6mbg5mX7Dg58fRuwe7TRJEnr4Z2iktQIC12SGmGhS1IjLHRJaoSFLkmNsNAlqREWurReziOuKdXpfeiS+pbmEV+aenZpHnHwrktNnGfo0no4j7immIUurYfziGuKWejSejiPuKaYhS6th/OIa4pZ6NJ67NsHc3Owcyckvc9zc74gqqngu1yk9XIecU0pz9AlqREWuiQ1wkKXpEZY6JLUCAtdkhphoUtSIyx0SWqEhS5JjbDQJakRFrokNaJToSfZm+S+JMeSXLPC8ecm+XSSO5McTXLl6KNKktYytNCTbAGuAy4B9gCXJ9mzbNhbgXur6uXAxcCfJjlrxFklSWvocoZ+EXCsqu6vqseAG4FLl40p4AeTBHg28ChwcqRJJUlr6lLo5wIPDWwv9PcN+iDwEuBh4G7g7VX1xPIHSrI/yXyS+cXFxWcYWZK0ki6FnhX21bLt1wFHgB8BLgQ+mOQ5T/tHVXNVNVtVszMzM+uMKklaS5dCXwDOG9jeTu9MfNCVwMer5xjwAHDBaCJKkrroUuiHgd1Jzu+/0HkZcNOyMQ8CrwFI8gLgxcD9owwqSVrb0BWLqupkkquBW4AtwPVVdTTJVf3jB4H3AR9Jcje9SzTvrqqvjzG3JGmZTkvQVdXNwM3L9h0c+Pph4JdGG02StB7eKSpJjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljoktQIC12SGmGhS1IjLHRJaoSFLkmNsNAlqREWuiQ1wkKXpEZY6JLUCAtdkhphoUtSIyx0SWqEhS5JjehU6En2JrkvybEk16xw/F1JjvQ/7knyeJLnjT6uJGk1Qws9yRbgOuASYA9weZI9g2Oq6gNVdWFVXQj8IfAvVfXoGPJKklbR5Qz9IuBYVd1fVY8BNwKXrjH+cuCGUYSTJHXXpdDPBR4a2F7o73uaJFuBvcDHVjm+P8l8kvnFxcX1ZpUkraFLoWeFfbXK2DcAX1jtcktVzVXVbFXNzszMdM0oSeqgS6EvAOcNbG8HHl5l7GV4uUWSJqJLoR8Gdic5P8lZ9Er7puWDkjwX+HngU6ONKEnq4sxhA6rqZJKrgVuALcD1VXU0yVX94wf7Q98I/ENVfXtsaSVJq0rVapfDx2t2drbm5+cn8tySdLpKckdVza50zDtFJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljoktQIC12SGmGhS1IjLHRJaoSFLkmNsNAlqREWuiQ1wkKXpEZY6JLUCAtdkhphoUtSIyx0SWqEhS5JjehU6En2JrkvybEk16wy5uIkR5IcTfIvo40pSRrmzGEDkmwBrgNeCywAh5PcVFX3Dow5G/gQsLeqHkzy/DHllSStossZ+kXAsaq6v6oeA24ELl025reAj1fVgwBV9choY0qShulS6OcCDw1sL/T3DXoRcE6Sf05yR5IrRhVQktTN0EsuQFbYVys8zk8CrwG+H7gtye1V9dWnPFCyH9gPsGPHjvWnlSStqssZ+gJw3sD2duDhFcZ8tqq+XVVfB24FXr78gapqrqpmq2p2ZmbmmWaWJK2gS6EfBnYnOT/JWcBlwE3LxnwK+LkkZybZCvwU8OXRRpUkrWXoJZeqOpnkauAWYAtwfVUdTXJV//jBqvpyks8CdwFPAB+uqnvGGVyS9FSpWn45fGPMzs7W/Pz8RJ5bkk5XSe6oqtmVjnmnqCQ1wkKXpEZY6JLUCAtdkhphoUtSIyx0SWqEhS5JjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljoktQIC12SGmGhS1IjLHRJaoSFLkmNsNAlqREWuiQ1olOhJ9mb5L4kx5Jcs8Lxi5N8I8mR/scfjT4qcOgQ7NoFZ5zR+3zo0FieRpJOR2cOG5BkC3Ad8FpgATic5KaqunfZ0M9X1S+PIWPPoUOwfz+cONHbPn68tw2wb9/YnlaSThddztAvAo5V1f1V9RhwI3DpeGOt4NprnyzzJSdO9PZLkjoV+rnAQwPbC/19y/10kjuTfCbJT6z0QEn2J5lPMr+4uLi+pA8+uL79krTJdCn0rLCvlm3/O7Czql4O/CXwyZUeqKrmqmq2qmZnZmbWFZQdO9a3X5I2mS6FvgCcN7C9HXh4cEBVfbOqvtX/+mbgWUm2jSwlwIEDsHXrU/dt3drbL0nqVOiHgd1Jzk9yFnAZcNPggCQvTJL+1xf1H/e/R5p03z6Ym4OdOyHpfZ6b8wVRSeob+i6XqjqZ5GrgFmALcH1VHU1yVf/4QeBNwO8lOQn8H3BZVS2/LHPq9u2zwCVpFRlH73YxOztb8/PzE3luSTpdJbmjqmZXOuadopLUCAtdkhphoUtSIyx0SWrExF4UTbIIHJ/Ik69tG/D1SYdYwTTmmsZMYK71Mtf6TDrXzqpa8c7MiRX6tEoyv9oryJM0jbmmMROYa73MtT7Tmgu85CJJzbDQJakRFvrTzU06wCqmMdc0ZgJzrZe51mdac3kNXZJa4Rm6JDXCQpekRmzKQu+w6PW7Bha8vifJ40meNwW5npvk0/2VoY4muXLcmTrmOifJJ5LcleSLSV66AZmuT/JIkntWOZ4kf9HPfFeSV4w7U8dcFyS5Lcl3krxzIzJ1zLWv/326K8m/Jnn5lOS6tJ/pSH+1s5+dhlwD417Z74c3bUSuoapqU33QmwL4P4AfBc4C7gT2rDH+DcA/TkMu4D3AH/e/ngEeBc6aglwfAN7b//oC4HMb8P16NfAK4J5Vjr8e+Ay9FbdeBfzbBv3/NSzX84FXAgeAd25Epo65fgY4p//1JVP0/Xo2T77W9zLgK9OQqz9mC/CPwM3AmzbqZ7nWx2Y8Q1/voteXAzdMSa4CfrC/mMiz6RX6ySnItQf4HEBVfQXYleQF4wxVVbfS++9fzaXAR6vnduDsJD88zkxdclXVI1V1GPjuuLMse95huf61qv6nv3k7vZXJpiHXt6rfnsAP8PTlLyeSq+9twMeAR8afqJvNWOhdF70myVZgL70f2jTk+iDwEnpLAN4NvL2qnpiCXHcCvwrfW7FqJxtUCGvo/HPW07yZ3l83UyHJG5N8Bfh74HcmnQcgybnAG4GDk84yaDMWepdFr5e8AfhCVQ37TT0KXXK9DjgC/AhwIfDBJM8Zb6xOud4PnJPkCL2zli8x/r8chlnPz1l9SX6BXqG/e9JZllTVJ6rqAuBXgPdNOM6SPwPeXVWPTzrIoKFL0DVo6KLXAy5jYy63QLdcVwLv7/8JeizJA/SuWX9xkrmq6pv9bPQvBz3Q/5ik9fycBSR5GfBh4JKqGu2awCNQVbcm+bEk26pq0pN2zQI39pdS3ga8PsnJqvrkJENtxjP0oYteQ+8dJcDPA5+aolwPAq/p53sB8GLg/knnSnJ2/xjAW4Bb+yU/STcBV/Tf7fIq4BtV9bUJZ5paSXYAHwd+u6q+Ouk8S5L8+MAC9K+g98L8xH/ZVNX5VbWrqnYBfwf8/qTLHDbhGXp1W/QaetfH/qGqvj1Fud4HfCTJ3fQuKbx73GcqHXO9BPhokseBe+n9yT5WSW4ALga2JVkA3gs8ayDTzfTe6XIMOEH/L4hJ50ryQmAeeA7wRJJ30HvX0Fh/AXb4fv0R8EPAh/r9ebI2YEbBDrl+jd4v5u/SW4D+NwdeJJ1krqnkrf+S1IjNeMlFkppkoUtSIyx0SWqEhS5JjbDQJWkDdJ3wa2D8byS5tz8R3193+je+y0WSxi/Jq4Fv0ZtjaM0ZSZPsBv4G+MWq+p8kz6+qoXPGeIYuSRtgpQm/+ne+fjbJHUk+n+SC/qHfBa5bmjCtS5mDhS5JkzQHvK2qfhJ4J/Ch/v4XAS9K8oUktyfZ2+XBNt2dopI0DZI8m9489H/bvzsX4Pv6n88EdtO7W3U78PkkL62q/13rMS10SZqMM4D/raoLVzi2ANxeVd8FHkhyH72CPzzsASVJG6w/f88DSX4dvrds4tLSf58EfqG/fxu9SzBDJ+Kz0CVpA/Qn/LoNeHGShSRvBvYBb05yJ3CUJ1cDuwX47yT3Av8EvKvLlMa+bVGSGuEZuiQ1wkKXpEZY6JLUCAtdkhphoUtSIyx0SWqEhS5Jjfh/3SkYqXSh2WgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "start=time.time()\n",
    "bwt_T=BWT(genome[0].seq.upper(),suffix_array_0)\n",
    "t0=time.time()\n",
    "bwt_T=BWT(genome[1].seq.upper(),suffix_array_1)\n",
    "t1=time.time()\n",
    "bwt_T=BWT(genome[2].seq.upper(),suffix_array_2)\n",
    "t2=time.time()\n",
    "bwt_T=BWT(genome[3].seq.upper(),suffix_array_3)\n",
    "t3=time.time()\n",
    "bwt_T=BWT(genome[4].seq.upper(),suffix_array_4)\n",
    "t4=time.time()\n",
    "bwt_T=BWT(genome[5].seq.upper(),suffix_array_5)\n",
    "t5=time.time()\n",
    "bwt_T=BWT(genome[6].seq.upper(),suffix_array_6)\n",
    "end=time.time()\n",
    "plt.plot([len(genome[0].seq.upper()), len(genome[1].seq.upper()), len(genome[2].seq.upper()),len(genome[3].seq.upper()), len(genome[4].seq.upper()), len(genome[5].seq.upper()), len(genome[6].seq.upper())], [t0-start,t1-t0, t2-t1, t3-t2,t4-t3,t5-t4, end-t5], 'ro')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27136fcc",
   "metadata": {},
   "source": [
    "On observe bien une fonction de complexité temporelle linéaire"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7e607d3",
   "metadata": {},
   "source": [
    "## 4/ Cherchons un pattern dans une séquence grâce à la transformée de Burrow-Wheeler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "4af714ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def new_suffix_table_positions_faster(BWT,pattern_letter, position_in_BWT, occurence_index, counts):\n",
    "    '''\n",
    "    Takes a list (parameter \"position_in_BWT\") of positions of letters in the BWT, \n",
    "    checks if those letters are equal to \"pattern_letter\", \n",
    "    and returns the positions in the BWT \n",
    "    of the letters that came right before those in \"position_in_BWT\" thats passed the check.\n",
    "    \n",
    "    Args:\n",
    "        BWT (list of string): Burrows-Wheeler transform of a sequence\n",
    "        pattern_letter (string): letter we want the elements of \"position_in_BWT\" to be equal to\n",
    "        position_in_BWT (list of int): list of positions of elements of the BWT to be studied\n",
    "        occurence_index (list of int): output of the function \"occurence_indexer\"\n",
    "        counts (dict): keys are each distinct letter of BWT, \n",
    "                       elements are how many times each letter appears\n",
    "    \n",
    "    Return:\n",
    "        output (list of int): new positions of elements of the BWT we want to study\n",
    "    '''\n",
    "    # Counter({'T': 3, 'A': 2, 'B': 2, 'C': 2, '!': 1})\n",
    "    output=[]\n",
    "    letter_and_occurence=[]\n",
    "    for i in range(len(position_in_BWT)):\n",
    "        if BWT[position_in_BWT[i]]==pattern_letter:\n",
    "            letter_and_occurence=[BWT[position_in_BWT[i]],occurence_index[position_in_BWT[i]]]\n",
    "            letter=letter_and_occurence[0]\n",
    "            occurence=letter_and_occurence[1]\n",
    "            new_position_to_evaluate=occurence + sum([counts[char] for char in counts if char < letter])\n",
    "            output.append(new_position_to_evaluate)\n",
    "    return output\n",
    "\n",
    "def initialize_suffix_table_positions(counts, pattern_letter):\n",
    "    '''\n",
    "    Finds the all the positions in the suffix table, of the letter \"pattern_letter\"\n",
    "    \n",
    "    Args:\n",
    "        counts (dict): keys are each distinct letter of BWT, \n",
    "                       elements are how many times each letter appears\n",
    "        pattern_letter (string): letter we are searching positions for\n",
    "    \n",
    "    Return:\n",
    "        (list of int): all the positions occupied by the letter \"pattern_letter\"\n",
    "                       in the suffix table\n",
    "    '''\n",
    "    start=sum([counts[char] for char in counts if char < pattern_letter])\n",
    "    end=start+counts[pattern_letter]\n",
    "    return [i for i in range(start, end)]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5fc872b",
   "metadata": {},
   "source": [
    "##### Complexité  \"new_suffix_table_positions_faster\": \n",
    "La fonction \"new_suffix_table_positions_faster\" parcourt le paramètre 'position_in_BWT' et dans la compréhension de liste, elle parcourt le paramètre 'counts'. Dans le cas où la séquence est un chromosome, 'counts' a une taille de 5. Elle a donc une complexité de $0(n+5)$ avec n la taille de 'position_in_BWT'. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4ce4c0a3",
   "metadata": {},
   "source": [
    "##### Complexité  de \"initialize_suffix_table_positions\": \n",
    "La fonction parcourt deux fois le paramètre \"counts\" (de taille 5 dans le cas où la séquence est un chromosome), somme chaque élément de cette liste (liste qui est donc de taille 4 dans le pire des cas), et termine en parcourant la différence entre les variables \"start\" et \"end\" . C'est de la taille de la somme de tous les éléments de \"counts\".\n",
    "\n",
    "Dans le cas où on travaille avec un chromosome de notre génome, elle est donc de complexité $0(n)$, avec n=taille du chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "273247d4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def occurrence_indexer(S):\n",
    "    '''\n",
    "    Returns a list of size len(S), from which each element i is the number of time S[i]\n",
    "    has appeared in the sub-sequence S[:i-1].\n",
    "    \n",
    "    Args:\n",
    "        S (list of string): sequence from which we want to count occurences\n",
    "    \n",
    "    Return:\n",
    "        K (list of int): occurence of each element of S\n",
    "    '''\n",
    "    K = []\n",
    "    last_index = {}\n",
    "    for s in S:\n",
    "        if s not in last_index: \n",
    "            last_index[s] = 0\n",
    "        K.append(last_index[s])\n",
    "        last_index[s] += 1\n",
    "    return(K)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f70aea66",
   "metadata": {},
   "source": [
    "##### Complexité  de \"occurrence_indexer\": \n",
    "Elle est de complexité $0(n)$, avec n=taille de S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "8b52c069",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_patterns_in_sequence_faster(sf, BWT, pattern):\n",
    "    '''\n",
    "        Aligns \"pattern\" in the sequence (from which we calculated \n",
    "        the parameters \"BWT\" and the suffix table \"sf\"). Returns all the \n",
    "        starting positions of the alignments found.\n",
    "\n",
    "    Args:\n",
    "        sf (list of int): suffix table of the sequence we want to find pattern in\n",
    "        BWT (list of string): BWT of the sequence we want to find pattern in\n",
    "        pattern (string): pattern we want to find in the sequence\n",
    "    \n",
    "    Return:\n",
    "        output (list of int): starting positions of all the alignments found\n",
    "    '''\n",
    "    output=[]\n",
    "    occurence_index = occurrence_indexer(BWT)\n",
    "    counts = Counter(BWT)\n",
    "    index=len(pattern)-1\n",
    "    pattern_letter=pattern[index]\n",
    "    \n",
    "    suffix_table_positions=initialize_suffix_table_positions(counts, pattern_letter)\n",
    "    while index>0 and len(suffix_table_positions)>0:\n",
    "        index-=1\n",
    "        pattern_letter=pattern[index]\n",
    "        suffix_table_positions=new_suffix_table_positions_faster(BWT,pattern_letter,suffix_table_positions,occurence_index,counts) \n",
    "        \n",
    "    for i in range(len(suffix_table_positions)):\n",
    "        output.append(sf[suffix_table_positions[i]])\n",
    "        \n",
    "    return output\n",
    "\n",
    "def find_patterns_in_sequence_faster_with_dc3(sequence, pattern, sequence_has_end_character=False):\n",
    "    '''\n",
    "        Aligns \"pattern\" in the sequence by calculating it's suffix table\n",
    "        and BWT. Returns all the \n",
    "        starting positions of the alignments found.\n",
    "\n",
    "    Args:\n",
    "        sequence (string): sequence where we want to align \"pattern\"\n",
    "        sequence_has_end_character (bool): if \"True\", we add \"!\" at the end of \"sequence\"\n",
    "        pattern (string): pattern we want to find in the sequence\n",
    "    \n",
    "    Return:\n",
    "        output (list of int): starting positions of all the alignments found\n",
    "    '''\n",
    "    output=[]\n",
    "    # adding an end character\n",
    "    if sequence_has_end_character==False:\n",
    "        sequence += '!'\n",
    "    \n",
    "    # computing the dc3\n",
    "    sf=almost_dc3_with_p12(asciiDC3(sequence)) \n",
    "    \n",
    "    #Computing the BWT\n",
    "    BWT = \"\"\n",
    "    for i in sf:\n",
    "        BWT += sequence[i-1]\n",
    "    \n",
    "    output=[]\n",
    "    occurence_index = occurrence_indexer(BWT)\n",
    "    counts = Counter(BWT)\n",
    "    index=len(pattern)-1\n",
    "    pattern_letter=pattern[index]\n",
    "    \n",
    "    suffix_table_positions=initialize_suffix_table_positions(counts, pattern_letter)\n",
    "    while index>0 and len(suffix_table_positions)>0:\n",
    "        index-=1\n",
    "        pattern_letter=pattern[index]\n",
    "        suffix_table_positions=new_suffix_table_positions_faster(BWT,pattern_letter,suffix_table_positions,occurence_index,counts) \n",
    "        \n",
    "    for i in range(len(suffix_table_positions)):\n",
    "        output.append(sf[suffix_table_positions[i]])\n",
    "        \n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "68e92f22",
   "metadata": {},
   "source": [
    "##### Complexité  de \"find_patterns_in_sequence_faster\": \n",
    " Les fonctions \"Counter\", \"occurence_indexer\", \"initialize_suffix_table_positions\", sont de complexité $0(n)$, avec n= taille de la séquence dans laquelle on cherche le pattern.\n",
    " \n",
    " La boucle while parcourt le paramètre \"pattern\" dont la taille est au pire des cas égale à n. Dans la boucle while, on trouve \"new_suffix_table_positions_faster\" de complexité $0(n)$.\n",
    " La boucle for parcourt toutes les positions d'alignement trouvées par la fonction, donc elle parcourt au pire n éléments.\n",
    " \n",
    " La fonction est donc de complexité $0(n²)$.\n",
    "."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0621a42a",
   "metadata": {},
   "source": [
    "##### Complexité  de \"find_patterns_in_sequence_faster_with_dc3\": \n",
    "La fonction diffère de find_patterns_in_sequence_faster par le fait qu'elle calcule la suffix table et la BWT de la séquence dans laquelle on veut aligner le pattern.\n",
    "\n",
    "Ces deux fonctions sont linéaires, la complexité reste $0(n²)$."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0ca1529c",
   "metadata": {},
   "source": [
    "### Cherchons un pattern simple dans un texte simple"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "efbbfe6e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Le pattern est à la positions [22, 6] du texte simple\n",
      "A la position 22 du texte simple, la séquence est GHIJKL\n",
      "A la position 6 du texte simple, la séquence est GHIJKL\n"
     ]
    }
   ],
   "source": [
    "texte_simple='ABCDEFGHIJKLMNOPABCDEFGHIJKLMNOP!'\n",
    "pattern_simple='GHIJK'\n",
    "\n",
    "#Calculons la suffix table du texte \n",
    "sf=almost_dc3_with_p12(asciiDC3(texte_simple))\n",
    "\n",
    "# Alignons le pattern dans le texte\n",
    "positions_of_pattern=find_patterns_in_sequence_faster_with_dc3(texte_simple, pattern_simple, sequence_has_end_character=True)\n",
    "print(\"Le pattern est à la positions \"+str(positions_of_pattern)+\" du texte simple\")\n",
    "\n",
    "# Montrons que l'alignement est juste\n",
    "for p in positions_of_pattern:\n",
    "    print(\"A la position \"+str(p)+\" du texte simple, la séquence est \"+str(texte_simple[p:p+6]))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f90fab65",
   "metadata": {},
   "source": [
    "## 5/  Importation des données expérimentales"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d641c437",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Importation du fichier BAM\n",
    "bam=pysam.AlignmentFile(\"/home/azarkua/Documents/2023-2024/omiques2/developement/omique2/single_Pfal_dat.bam\", \"rb\")\n",
    "#bam = pysam.AlignmentFile(\"/home/mvernier/Documents/2023-2024/omique2/projet/single_Pfal_dat.bam\", \"rb\")\n",
    "\n",
    "donnees_experimentales=[]\n",
    "for alignment in bam:\n",
    "    read_name = alignment.query_name\n",
    "    reference_name = bam.getrname(alignment.reference_id)\n",
    "    position = alignment.reference_start\n",
    "    mapping_quality = alignment.mapping_quality\n",
    "    \n",
    "    donnees_experimentales.append(position)\n",
    "    \n",
    "bam.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e058481",
   "metadata": {},
   "source": [
    "## 6/ Créer et aligner les kmers de nos reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e6cadc01",
   "metadata": {},
   "outputs": [],
   "source": [
    "def kmer(sequence, longueur):\n",
    "    '''\n",
    "        Creates kmers of size \"longueur\" from \"sequence\"\n",
    "\n",
    "    Args:\n",
    "        sequence (string): sequence we want to create kmers from\n",
    "        longueur (int): size of kmers\n",
    "    \n",
    "    Return:\n",
    "        kmer (list of string): list of all the kmers\n",
    "    '''\n",
    "    n = len(sequence)\n",
    "    kmer=[]\n",
    "    for i in range(n-longueur+1):\n",
    "        kmer.append(sequence[i:i+longueur])\n",
    "    return kmer\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38ca117f",
   "metadata": {},
   "source": [
    "##### Complexité : \n",
    "kmer parcourt 'sequence'-longueur+1, elle a donc une complexité de $0(n)$ avec n=taille de 'sequence'-longueur+1"
   ]
  },
  {
   "cell_type": "raw",
   "id": "79c36408",
   "metadata": {},
   "source": [
    "def position_with_suffix_table(liste_kmer, genome):\n",
    "\n",
    "    output=[]\n",
    "    for i in range(len(liste_kmer)):\n",
    "        output.append(find_patterns_in_sequence_faster_with_dc3(genom, liste_kmer[i],sequence_has_end_character=True))\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6b31e2ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "def position(liste_kmer, bwt, suffix_table):\n",
    "    '''\n",
    "        Return all the position that each kmer takes in the genome.\n",
    "    Args : \n",
    "        liste_kmer (list of string) : list of kmer\n",
    "        bwt(string): burrow wheeler transform of the genome in which we want to find the kmers \n",
    "        suffix_table(list of int): suffix table of the genome in which we want to find the kmers \n",
    "    Return : \n",
    "        output (list of list of int) : list of all the position that each kmers takes in the genome \n",
    "    '''\n",
    "    output=[]\n",
    "    for i in range(len(liste_kmer)):\n",
    "        output.append(find_patterns_in_sequence_faster(suffix_table, bwt, liste_kmer[i]))\n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79ee3877",
   "metadata": {},
   "source": [
    "##### Complexité : \n",
    "position_with_suffix_table parcourt 'liste_kmer' et utilise 'find_patterns_in_sequence_with_dc3' (qui a une complexité de ?), elle a donc une complexité de $0()$\n",
    "\n",
    "position parcourt 'liste_kmer' et utilise 'find_patterns_in_sequence_with_dc3' (qui a une complexité de ?), elle a donc une complexité de $0()$\n",
    "\n",
    "position_sorted parcourt 'liste_kmer' et utilise 'find_patterns_in_sequence_with_dc3' (qui a une complexité de ?) et 'sorted' (qui a une complexité $0(nlogn)$, n étant la taille de la séquence fournie), elle a donc une complexité de $0()$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "7397f269",
   "metadata": {},
   "outputs": [],
   "source": [
    "def complementaire_inverse(sequence):\n",
    "    '''\n",
    "        The function return the inverse complementary of the sequence \n",
    "    \n",
    "    Args : \n",
    "        sequence(String): sequence \n",
    "        \n",
    "    Return : \n",
    "        ''.join(ci) (String): inverse complementary of the sequence \n",
    "    '''\n",
    "    ci = []\n",
    "    for i in range(len(sequence) - 1, -1, -1):\n",
    "        if sequence[i] == 'A':\n",
    "            ci.append('T')\n",
    "        elif sequence[i] == 'T':\n",
    "            ci.append('A')\n",
    "        elif sequence[i] == 'C':\n",
    "            ci.append('G')\n",
    "        elif sequence[i] == 'G':\n",
    "            ci.append('C')\n",
    "    return ''.join(ci)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7daaceb",
   "metadata": {},
   "source": [
    "##### Complexité : \n",
    "complementaire_inverse parcourt 'sequence' et utilise 'join' qui a une complexité linéaire,  elle a donc une complexité $0(n²)$, n étant la taille de 'sequence'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c978427e",
   "metadata": {},
   "source": [
    "## Première approche d'alignement: rapide mais qui ne gère pas les cas complexes \n",
    "Nos kmers sont de taille 10.\n",
    "On aligne les kmers du débuts et de la fin du read.\n",
    "Puis on assemble les kmers dont les alignements sont espacés de la taille du read ( plus ou moins un taux d'erreur à fixer).\n",
    "\n",
    "Dans le cas où le read a une mutation/ déletion/ insertion dans ses nucléotides initiaux ou finaux, on prends les kmers du début et de la fin, décalés d'une longueur de 10.\n",
    "\n",
    "Dans le cas où aucun alignement n'est trouvé, on recommence avec le complémentaire inverse du génome. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "bbfe987e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assembler_premier_kmer(positions, nb_de_kmer_a_aligner):\n",
    "    \"\"\"\n",
    "    Return the positions of kmer in position 1  in the genome that have passed the following test : \n",
    "    we check if the following kmer are aligned \n",
    "    \n",
    "    \n",
    "    Example : \n",
    "    position taken by first kmer  [0, 20, 50]\n",
    "    position taken by second kmer [1, 21] ---> the algorithm removes \"50\" from the starting positions\n",
    "    position taken by third kmer  [2] -------> the algorithm removes \"20 from the starting positions\"\n",
    "    \n",
    "    assembler_premier_kmer(positions, 3)=[0]\n",
    "    \n",
    "    Beware : The positions needs to be sorted for this function to work\n",
    "    \n",
    "    Args : \n",
    "        positions(list of list of int) : list of all the positions of the kmers in the genome \n",
    "        nb_kmer_a_aligner(int): number of kmers we want to be aligned \n",
    "        \n",
    "    Return : \n",
    "        positions_initiales(list of list of int): the positions of kmer in position 1  in the genome that \n",
    "                                                  have passed the test\n",
    "    \"\"\"\n",
    "\n",
    "    positions_initiales=positions[0]\n",
    "\n",
    "    for i in range(1,nb_de_kmer_a_aligner):\n",
    "        current_kmer_positions=positions[i]\n",
    "        index_1=0\n",
    "        index_2=0\n",
    "        new_initial_positions=copy.deepcopy(positions_initiales)\n",
    "        \n",
    "        while index_1<len(positions_initiales) and index_2<len(current_kmer_positions):\n",
    "\n",
    "            while positions_initiales[index_1]+i>current_kmer_positions[index_2] and index_2+1<len(current_kmer_positions):\n",
    "                index_2+=1\n",
    "            if positions_initiales[index_1]+i!=current_kmer_positions[index_2]:\n",
    "                new_initial_positions.remove(positions_initiales[index_1])\n",
    "                index_1+=1\n",
    "            else:\n",
    "\n",
    "                index_1+=1\n",
    "                \n",
    "        positions_initiales=new_initial_positions\n",
    "                     \n",
    "    return positions_initiales\n",
    "\n",
    "def assembler_dernier_kmer(positions, nb_de_kmer_a_aligner):\n",
    "    \"\"\"\n",
    "\n",
    "    This function is the same as 'assembler_premier_kmer' but starting from the last kmer and \n",
    "    we want the previous kmers to be aligned \n",
    "    \n",
    "    Example:\n",
    "    position taken by last kmer        [2, 20, 50]\n",
    "    position taken by second last kmer [1, 19] ---> the algorithm removes \"50\" from the final positions\n",
    "    position taken by third last kmer  [0] -------> the algorithm removes \"20 from the final positions\"\n",
    "    \n",
    "    assembler_premier_kmer(positions, 3)=[2]\n",
    "    \n",
    "    Beware : The positions needs to be sorted for this function to work\n",
    "    \n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    positions_finales=positions[-1]\n",
    "    for i in range(1,nb_de_kmer_a_aligner):\n",
    "        current_kmer_positions=positions[-1-i]\n",
    "        index_1=len(positions_finales)-1\n",
    "        index_2=len(current_kmer_positions)-1\n",
    "        new_final_positions=copy.deepcopy(positions_finales)\n",
    "        \n",
    "        while index_1>=0 and index_2>=0:\n",
    "\n",
    "            while positions_finales[index_1]-i<current_kmer_positions[index_2] and index_2>=0:\n",
    "                index_2-=1\n",
    "            if positions_finales[index_1]-i!=current_kmer_positions[index_2]:\n",
    "                new_final_positions.remove(positions_finales[index_1])\n",
    "                index_1-=1\n",
    "            else:\n",
    "\n",
    "                index_1-=1\n",
    "                \n",
    "        positions_finales=new_final_positions\n",
    "                     \n",
    "    return positions_finales"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "11b984d6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def aligner_un_read_faster(read, taille_kmer, positions, nb_de_kmer_a_aligner, erreurs, decalage):\n",
    "    '''\n",
    "    Finds the positions in the first kmer that are at the distance of the size of the read \n",
    "    (taking a given amount of mistakes into account: of value parameter \"erreurs\") of the finals kmers. \n",
    "    It can take into account a given gap (parameter \"decalage\")\n",
    "    \n",
    "    Args : \n",
    "        read(string): read\n",
    "        taille_kmer(int): size of the kmers\n",
    "        positions(list of list of int): positions of the kmers in the genome \n",
    "        nb_de_kmer_a_aligner(int): number of kmer we want to aligne with the functions 'assembler_premier_kmer'\n",
    "                                   and 'assembler_dernier_kmer'\n",
    "        erreurs(int): number of mistakes we allow\n",
    "        decalage(int): the \"initial kmer\" and \"final kmer\" have a gap of size \"decalage\"\n",
    "                       from the first and last kmer of our read.\n",
    "    Return : \n",
    "        output(list of int): all the positions of the fist kmer at the correct distance from the last kmer\n",
    "    '''\n",
    "    positions_initiales=assembler_premier_kmer(positions, nb_de_kmer_a_aligner)\n",
    "    positions_finales=assembler_dernier_kmer(positions, nb_de_kmer_a_aligner)\n",
    "\n",
    "    output=[]\n",
    "    for i in range(len(positions_initiales)):\n",
    "        one_align=0\n",
    "        for j in range(len(positions_finales)):\n",
    "            if positions_initiales[i]+(len(read)-2*decalage-taille_kmer)-erreurs<=positions_finales[j]<=positions_initiales[i]+(len(read)-2*decalage-taille_kmer)+erreurs:\n",
    "                one_align+=1\n",
    "        if one_align>0:\n",
    "            output.append(positions_initiales[i])\n",
    "            \n",
    "    return output\n",
    "\n",
    "def aligner_un_read_avec_mutation(read, taille_kmer, liste_kmer, bwt, suffix_table, comp_inv_bwt, comp_inv_suffix_table,nb_de_kmer_a_aligner, erreurs, complementaire=False):\n",
    "    '''\n",
    "    \n",
    "    \n",
    "    '''\n",
    "    if complementaire:\n",
    "        BWT=comp_inv_bwt\n",
    "        sf=comp_inv_suffix_table\n",
    "    else:\n",
    "        BWT=bwt\n",
    "        sf=suffix_table\n",
    "    positions= position_partial_sorted(liste_kmer, BWT, sf, nb_de_kmer_a_aligner, 0)\n",
    "    output=aligner_un_read_faster(read, taille_kmer, positions, nb_de_kmer_a_aligner, erreurs,0)\n",
    "    if len(output)!=1:\n",
    "        # si on obtient pas la position initiale à partir du premier et dernier kmer, c'est probablement dû à une\n",
    "        # mutation sur les premiers/derniers nucléotides du read\n",
    "        \n",
    "        #on décale nos recherches sur le 10eme premier kmer et 10e kmer avant la fin.\n",
    "        positions= position_partial_sorted(liste_kmer, BWT, sf, nb_de_kmer_a_aligner, 10)\n",
    "        output=aligner_un_read_faster(read, taille_kmer, positions, nb_de_kmer_a_aligner, erreurs, 10)\n",
    "        # si rien ne marche, on essaie d'aligner au complémentaire\n",
    "        if len(output)!=1 and complementaire==False:\n",
    "            output=aligner_un_read_avec_mutation(read, taille_kmer, liste_kmer, bwt, suffix_table, comp_inv_bwt, comp_inv_suffix_table, nb_de_kmer_a_aligner, erreurs, complementaire=True)\n",
    "        elif len(output)!=1 and complementaire==True:\n",
    "            return [-1]\n",
    "        else:\n",
    "            output[0]-=taille_kmer\n",
    "    return output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "804fd8e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def position_partial_sorted(liste_kmer, bwt, suffix_table, nb_kmer, decalage):\n",
    "    debut=[]\n",
    "    fin=[0 for i in range(nb_kmer)]\n",
    "    for i in range(nb_kmer):\n",
    "        debut.append(sorted(find_patterns_in_sequence_faster(suffix_table, bwt, liste_kmer[decalage+i])))\n",
    "        fin[-1-i]=sorted(find_patterns_in_sequence_faster(suffix_table, bwt, liste_kmer[-1-decalage-i]))\n",
    "    return debut+fin"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "39c20584",
   "metadata": {},
   "source": [
    "### Testons notre méthode d'alignement avec le premier read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "6d76f6e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_read=reads[1].seq.upper()\n",
    "current_genome=genome[0].seq.upper()\n",
    "comp_inv_current_genome=complementaire_inverse(genome[0].seq.upper())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "0fdc2b19",
   "metadata": {},
   "outputs": [],
   "source": [
    "sf=almost_dc3_with_p12(asciiDC3(current_genome+\"!\")) \n",
    "# suffix table du 1er chromosome\n",
    "bwt=BWT(current_genome+\"!\", sf, True)\n",
    "comp_inv_sf=almost_dc3_with_p12(asciiDC3(comp_inv_current_genome+\"!\"))\n",
    "# suffix table du complémentaire inverse du 1er chromosome\n",
    "comp_inv_bwt=BWT(comp_inv_current_genome+\"!\", sf, True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "936cc4cc",
   "metadata": {},
   "source": [
    "Calcul de tous les kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "9500fb94",
   "metadata": {},
   "outputs": [],
   "source": [
    "liste_kmer=kmer(current_read, 10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aad02bc1",
   "metadata": {},
   "source": [
    "Alignement de notre read"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "804ebb65",
   "metadata": {},
   "outputs": [],
   "source": [
    "position_initiale_read=aligner_un_read_avec_mutation(current_read, 10, liste_kmer, bwt,sf, comp_inv_bwt, comp_inv_sf, 3, 2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "ca32f2aa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "position initiale proposée par l'algorithme\n",
      "143900\n",
      "Le read est: \n",
      "TATATCTTTAAAATGATGTTGCAAATTTATTGAACATGTTAATAAATCATCCTGTTCATTTTGTATGTCTACTAAATTATGTAACGTATCCTCTTCTTCA\n",
      "Les 100 nucléotides de notre genome à la position 143900 est:\n",
      "TATATCTTTAAAATGATGTTGCAAATTTATTGAACATGTTAATAAATCATCCTGTTCATTTTGTATGTCTACTAAATTATGTAACGTATCCTTTTCTTCA\n"
     ]
    }
   ],
   "source": [
    "print(\"position initiale proposée par l'algorithme\")\n",
    "print(position_initiale_read[0])\n",
    "print(\"Le read est: \")\n",
    "print(current_read)\n",
    "print(\"Les 100 nucléotides de notre genome à la position \"+str(position_initiale_read[0])+\" est:\")\n",
    "print(current_genome[position_initiale_read[0]:position_initiale_read[0]+len(current_read)])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba7d0232",
   "metadata": {},
   "source": [
    "Alignons 100 reads pour voir le taux d'erreur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "041d3853",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_36096/2796850528.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m100\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0mliste_kmer\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mkmer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreads\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m     \u001b[0malign_reads\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m{\u001b[0m\u001b[0;34m\"read\"\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m\"alignement\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0maligner_un_read_avec_mutation\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mreads\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mseq\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mupper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m10\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mliste_kmer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbwt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcomp_inv_bwt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcomp_inv_sf\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m3\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m}\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m \u001b[0mend\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtime\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Aligner 100 reads prend \"\u001b[0m\u001b[0;34m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mend\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mstart\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_36096/3948583188.py\u001b[0m in \u001b[0;36maligner_un_read_avec_mutation\u001b[0;34m(read, taille_kmer, liste_kmer, bwt, suffix_table, comp_inv_bwt, comp_inv_suffix_table, nb_de_kmer_a_aligner, erreurs, complementaire)\u001b[0m\n\u001b[1;32m     22\u001b[0m         \u001b[0mBWT\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mbwt\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     23\u001b[0m         \u001b[0msf\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msuffix_table\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 24\u001b[0;31m     \u001b[0mpositions\u001b[0m\u001b[0;34m=\u001b[0m \u001b[0mposition_partial_sorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mliste_kmer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mBWT\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnb_de_kmer_a_aligner\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     25\u001b[0m     \u001b[0moutput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maligner_un_read_faster\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtaille_kmer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpositions\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnb_de_kmer_a_aligner\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0merreurs\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0moutput\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_36096/2625266179.py\u001b[0m in \u001b[0;36mposition_partial_sorted\u001b[0;34m(liste_kmer, bwt, suffix_table, nb_kmer, decalage)\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mfin\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnb_kmer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnb_kmer\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m         \u001b[0mdebut\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfind_patterns_in_sequence_faster\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msuffix_table\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbwt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mliste_kmer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mdecalage\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m         \u001b[0mfin\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfind_patterns_in_sequence_faster\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msuffix_table\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mbwt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mliste_kmer\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mdecalage\u001b[0m\u001b[0;34m-\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mdebut\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mfin\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_36096/3910972299.py\u001b[0m in \u001b[0;36mfind_patterns_in_sequence_faster\u001b[0;34m(sf, BWT, pattern)\u001b[0m\n\u001b[1;32m     23\u001b[0m         \u001b[0mindex\u001b[0m\u001b[0;34m-=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m         \u001b[0mpattern_letter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mpattern\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 25\u001b[0;31m         \u001b[0msuffix_table_positions\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnew_suffix_table_positions_faster\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mBWT\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mpattern_letter\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msuffix_table_positions\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0moccurence_index\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mcounts\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     26\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msuffix_table_positions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/tmp/ipykernel_36096/4115671182.py\u001b[0m in \u001b[0;36mnew_suffix_table_positions_faster\u001b[0;34m(BWT, pattern_letter, position_in_BWT, occurence_index, counts)\u001b[0m\n\u001b[1;32m     25\u001b[0m             \u001b[0mletter\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mletter_and_occurence\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m             \u001b[0moccurence\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mletter_and_occurence\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 27\u001b[0;31m             \u001b[0mnew_position_to_evaluate\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0moccurence\u001b[0m \u001b[0;34m+\u001b[0m \u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcounts\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mchar\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mchar\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcounts\u001b[0m \u001b[0;32mif\u001b[0m \u001b[0mchar\u001b[0m \u001b[0;34m<\u001b[0m \u001b[0mletter\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     28\u001b[0m             \u001b[0moutput\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnew_position_to_evaluate\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "start=time.time()\n",
    "align_reads=[]\n",
    "for i in range(100):\n",
    "    liste_kmer=kmer(reads[i].seq.upper(), 10)\n",
    "    align_reads.append({\"read\": i, \"alignement\":aligner_un_read_avec_mutation(reads[i].seq.upper(), 10, liste_kmer, bwt, sf, comp_inv_bwt, comp_inv_sf,3, 2)})\n",
    "end=time.time()\n",
    "print(\"Aligner 100 reads prend \"+ str(end-start))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4bf4579",
   "metadata": {},
   "source": [
    "Les alignements proposés sont:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "83e7361d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'read': 0, 'alignement': [-1]}, {'read': 1, 'alignement': [143900]}, {'read': 2, 'alignement': [471737]}, {'read': 3, 'alignement': [65426]}, {'read': 4, 'alignement': [-1]}]\n"
     ]
    }
   ],
   "source": [
    "print(align_reads[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0d111cf",
   "metadata": {},
   "source": [
    "Calculons le taux d'erreur"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "bcd95cd1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Le pourcentage d'erreur est 53.84615384615385%\n"
     ]
    }
   ],
   "source": [
    "error=0\n",
    "for i in range(len(align_reads)):\n",
    "    if (donnees_experimentales[i]!=align_reads[i][\"alignement\"][0]):\n",
    "        error+=1\n",
    "pourcentage_erreur=(error/len(align_reads))*100\n",
    "print(\"Le pourcentage d'erreur est \"+str(pourcentage_erreur)+\"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92432832",
   "metadata": {},
   "source": [
    "## Seconde approche d'alignement: lente mais juste \n",
    "\n",
    "On assemble tous les kmers un par uns à la suite.   \n",
    "Dans le cas d'une mutation, on continue l'alignement à partir du dixième kmer plus loin.  \n",
    "Si aucun alignement n'est trouvé, on s'interesse au complémentaire inverse du read."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "6b313a23",
   "metadata": {},
   "outputs": [],
   "source": [
    "def assembler_des_kmers(numero_du_kmer, position_ds_genome,longueur, liste_position_kmers):\n",
    "    compteur_de_kmer=1 #compte le nb de kmers qu'on arrive à aligner\n",
    "    last_position=0\n",
    "    i=numero_du_kmer+1\n",
    "    difference=1\n",
    "    \n",
    "    while i<len(liste_position_kmers):\n",
    "        j=0\n",
    "        previous_compteur=compteur_de_kmer\n",
    "        k=0\n",
    "        while j<len(liste_position_kmers[i]) and compteur_de_kmer==previous_compteur:\n",
    "            # on ne quitte pas la boucle while tant qu'on a pas réussi\n",
    "            # à aligner un kmer, où qu'on a pas parcouru la liste des positions prises par le kmer\n",
    "            if position_ds_genome+difference==liste_position_kmers[i][j]:\n",
    "                compteur_de_kmer+=1\n",
    "                last_position=liste_position_kmers[i][j]\n",
    "            j+=1\n",
    "            \n",
    "        if compteur_de_kmer==previous_compteur:\n",
    "            # si on a pas réussi à aligner ce kmer\n",
    "            i+=longueur\n",
    "            # on continue l'alignement à partir du kmer 10 positions plus loin\n",
    "            difference+=longueur\n",
    "            if  i>= len(liste_position_kmers):\n",
    "                # on vérifie que 10 positions plus loin on aie pas atteint la fin de notre read\n",
    "                return [[position_ds_genome,last_position+longueur],compteur_de_kmer]\n",
    "            else:\n",
    "                \n",
    "                #on teste si l'absence d'alignement est du à un mutation\n",
    "                k=0\n",
    "\n",
    "                while k<len(liste_position_kmers[i]) and compteur_de_kmer==previous_compteur:\n",
    "                    if position_ds_genome+difference==liste_position_kmers[i][k]:\n",
    "                        compteur_de_kmer+=1\n",
    "                        last_position=liste_position_kmers[i][k]\n",
    "                    k+=1\n",
    "\n",
    "\n",
    "\n",
    "                if compteur_de_kmer==previous_compteur:\n",
    "                    ###cas d'une insertion\n",
    "                    difference-=1\n",
    "                    l=0\n",
    "                    while l<len(liste_position_kmers[i]) and compteur_de_kmer==previous_compteur:\n",
    "\n",
    "                        if position_ds_genome+difference==liste_position_kmers[i][l]:\n",
    "                            compteur_de_kmer+=1\n",
    "                            last_position=liste_position_kmers[i][l]\n",
    "                        l+=1\n",
    "\n",
    "                    if compteur_de_kmer==previous_compteur:\n",
    "                        ###cas d'une déletions\n",
    "                        difference+=1\n",
    "                        i-=1\n",
    "                        m=0\n",
    "                        while m<len(liste_position_kmers[i]) and compteur_de_kmer==previous_compteur:\n",
    "\n",
    "                            if position_ds_genome+difference==liste_position_kmers[i][m]:\n",
    "                                compteur_de_kmer+=1\n",
    "                                last_position=liste_position_kmers[i][m]\n",
    "                            m+=1\n",
    "\n",
    "                        if compteur_de_kmer==previous_compteur:\n",
    "                            ### cas où l'alignement est terminé\n",
    "                            return [[position_ds_genome,last_position+longueur],compteur_de_kmer]\n",
    "\n",
    "        i+=1\n",
    "        difference+=1   \n",
    "        \n",
    "    return [[position_ds_genome,last_position+longueur],compteur_de_kmer]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "b2ce986e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def alignement_maximum_de_kmer(liste_position_kmers, longueur):\n",
    "    output=[[0,0],0]\n",
    "    maximum=0\n",
    "    for i in range(longueur+1):\n",
    "        current_kmer=liste_position_kmers[i]\n",
    "        for p in current_kmer:\n",
    "            alignement=assembler_des_kmers(i, p, longueur, liste_position_kmers)\n",
    "            if maximum<alignement[1]:\n",
    "                output=alignement\n",
    "                maximum=alignement[1]\n",
    "    return output"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03d87bda",
   "metadata": {},
   "source": [
    "##### Complexité : \n",
    "alignement_maximum_de_kmer parcourt 'longueur'+1 et 'current_kmer' et utilise 'assembler_des_kmers'(qui a une complexité $0()$, elle a donc une complexité $0()$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "86062fd9",
   "metadata": {},
   "outputs": [],
   "source": [
    "def align_read(chromosome, reads_list, longueur_kmer):\n",
    "    sf=almost_dc3_with_p12(asciiDC3(chromosome))\n",
    "    bwt=BWT(chromosome, sf)\n",
    "    output=[]\n",
    "    \n",
    "    for i in range(len(reads_list)): # on parcourt la liste de toutes les positions des kmers\n",
    "        liste_de_kmer=kmer(reads_list[i].seq.upper(), longueur_kmer)\n",
    "        position=[]\n",
    "        for j in range(len(liste_de_kmer)):\n",
    "            # on démarre notre recherche d'alignement à partir des 10 premiers\n",
    "            # kmers, pour anticiper le cas où une SNP se trouve au début du read\n",
    "            position.append(find_patterns_in_sequence_faster(sf,bwt, liste_de_kmer[j]))\n",
    "        \n",
    "        alignment=alignement_maximum_de_kmer(position, longueur_kmer)\n",
    "        \n",
    "        \n",
    "        if (alignment[0][1]-alignment[0][0])<30:\n",
    "            # si l'alignement est trop mauvais\n",
    "            # on cherche à aligner le complémentaire inverse\n",
    "            liste_de_kmer=kmer(complementaire_inverse(reads_list[i].seq.upper()), longueur_kmer)\n",
    "            position=[]\n",
    "            for j in range(len(liste_de_kmer)):\n",
    "                position.append(find_patterns_in_sequence_faster(sf,bwt, liste_de_kmer[j]))\n",
    "                \n",
    "            alignment=alignement_maximum_de_kmer(position, longueur_kmer)\n",
    "            \n",
    "        # on crée un immense liste avec tous les alignements trouvés        \n",
    "        output.append({\"read\":i, \"alignement\":alignment})\n",
    "    return output\n",
    "        "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3f3a8750",
   "metadata": {},
   "source": [
    "##### Complexité :\n",
    "align_read parcourt 'reads_list', 'liste_kmer' utilise almost_dc3_with_p12, asciiDC3, BWT, kmer, find_patterns_in_sequence_faster, alignement_maximum_de_kmer et dans le pire des cas (si on doit chercher dans le complémentaire inverse) utilise aussi complementaire_inverse, find_patterns_in_sequence_faster, alignement_maximum_de_kmer et reparcourt liste_de_kmer. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "43776d37",
   "metadata": {},
   "source": [
    "### Testons notre algorithme sur le premier read:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b171914b",
   "metadata": {},
   "outputs": [],
   "source": [
    "premier_chr=genome[0].seq.upper()+\"!\"\n",
    "sf=almost_dc3_with_p12(asciiDC3(premier_chr)) # suffix table du 1er chromosome\n",
    "bwt=BWT(premier_chr, sf, True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "f7c42a89",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'read': 0, 'alignement': [[131734, 131834], 81]}, {'read': 1, 'alignement': [[143900, 143992], 83]}, {'read': 2, 'alignement': [[471737, 471837], 91]}]\n",
      "Aligner 5 reads prend 94.90097427368164\n"
     ]
    }
   ],
   "source": [
    "start=time.time()\n",
    "align=align_read(premier_chr,reads[:3], 10)\n",
    "end=time.time()\n",
    "print(align)\n",
    "print(\"Aligner 5 reads prend \"+str(end-start))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "49677892",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'read': 0, 'alignement': [[131734, 131834], 81]}, {'read': 1, 'alignement': [[143900, 143992], 83]}, {'read': 2, 'alignement': [[471737, 471837], 91]}]\n"
     ]
    }
   ],
   "source": [
    "print(align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "bba916e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[131734, 143900, 471737, 65426, 152677]\n"
     ]
    }
   ],
   "source": [
    "print(donnees_experimentales[:5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "id": "007ad7af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "131734\n",
      "131734\n",
      "143900\n",
      "143900\n",
      "471737\n",
      "471737\n",
      "Le pourcentage d'erreur est 0.0%\n"
     ]
    }
   ],
   "source": [
    "error=0\n",
    "for i in range(len(align)):\n",
    "    print(donnees_experimentales[i])\n",
    "    print(align[i][\"alignement\"][0][0])\n",
    "    if (donnees_experimentales[i]!=align[i][\"alignement\"][0][0]):\n",
    "        error+=1\n",
    "pourcentage_erreur=(error/len(align))*100\n",
    "print(\"Le pourcentage d'erreur est \"+str(pourcentage_erreur)+\"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a62d5460",
   "metadata": {},
   "source": [
    "### Visualisation des reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18649f8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_for_mutation(index, read_decalage,genome_decalage, read,genome):\n",
    "    return read[read_decalage+index+1]==genome[genome_decalage+index+1]\n",
    "\n",
    "\n",
    "def check_for_deletion(index, read_decalage, genome_decalage, read,genome):\n",
    "    return read[read_decalage+index]==genome[genome_decalage+index+1]\n",
    "\n",
    "def check_for_addition(index, read_decalage, genome_decalage, read,genome):\n",
    "    return read[read_decalage+index+1]==genome[genome_decalage+index]\n",
    "\n",
    "def compare_read_and_genome(position_in_genome, read,genome):\n",
    "    read_decalage=0\n",
    "    genome_decalage=position_in_genome\n",
    "    str_read=\"\"\n",
    "    str_genome=\"\"\n",
    "    i=0\n",
    "    while i+read_decalage<len(read):\n",
    "        if read[read_decalage+i]!=genome[genome_decalage+i]:\n",
    "            if check_for_mutation(i, read_decalage,genome_decalage, read,genome):\n",
    "                str_read+=\"*\"+read[read_decalage+i]+\"*\"\n",
    "                str_genome+=\"*\"+genome[genome_decalage+i]+\"*\"\n",
    "                i+=1\n",
    "                \n",
    "            elif check_for_deletion(i, read_decalage,genome_decalage, read,genome):\n",
    "                str_read+=\"-\"\n",
    "                str_genome+=genome[genome_decalage+i]\n",
    "                genome_decalage+=1\n",
    "                i+=1\n",
    "                \n",
    "            elif check_for_addition(i, read_decalage, genome_decalage, read,genome):\n",
    "                str_read+=read[read_decalage+i]\n",
    "                str_genome+=\"+\"+genome[genome_decalage+i]\n",
    "                i+=1\n",
    "                read_decalage+=1\n",
    "                \n",
    "            else:\n",
    "                i+=1\n",
    "                print(\"Nous observons plus d'une erreur d'alignement à la suite qui se suivent entre les positions \"+str(genome_decalage+i-1)+\" et \"+str(genome_decalage+i))\n",
    "        else:\n",
    "            str_read+=read[read_decalage+i]\n",
    "            str_genome+=genome[genome_decalage+i]\n",
    "            i+=1\n",
    "            \n",
    "    print(\"read\")\n",
    "    print(str_read)\n",
    "        \n",
    "    print(\"genome\")\n",
    "    print(str_genome)\n",
    "    return"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c9e46378",
   "metadata": {},
   "source": [
    "#### Testons notre algorithme sur un exemple "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "96ead068",
   "metadata": {},
   "outputs": [],
   "source": [
    "current_read=reads[1].seq.upper()\n",
    "current_genome=genome[0].seq.upper()\n",
    "starting_position=143900"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c43da50c",
   "metadata": {},
   "source": [
    "Les fonctions suivantes nous permettent de visualiser les zones de SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c68ba00",
   "metadata": {},
   "outputs": [],
   "source": [
    "compare_read_and_genome(starting_position,current_read,current_genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "91487c3d",
   "metadata": {},
   "outputs": [],
   "source": [
    "cProfile.run(\"align_read(premier_chr,reads[5:10], 10)\", \"pattern_matching_stats\")\n",
    "p = pstats.Stats(\"pattern_matching_stats\")\n",
    "p.sort_stats(\"cumulative\").print_stats()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60780f09",
   "metadata": {},
   "source": [
    "### Comparaison des résultats théoriques avec les résultats expérimentaux"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "747168a4",
   "metadata": {},
   "source": [
    "### Analyse de la qualité des séquences "
   ]
  },
  {
   "attachments": {
    "Capture%20d%E2%80%99%C3%A9cran%20du%202023-11-09%2015-44-16.png": {
     "image/png": "iVBORw0KGgoAAAANSUhEUgAAA4cAAAKTCAIAAAAlv+r4AAAAA3NCSVQICAjb4U/gAAAAGXRFWHRTb2Z0d2FyZQBnbm9tZS1zY3JlZW5zaG907wO/PgAAIABJREFUeJzt3X9wXNdhH/qzJCTAlmKuI7oEISVELKqEIzVEao8JN0IFx54Qmeg9AXYaIX3uCOlTI3riRkxfZ8zEbzrUqxrDHeeFmuc+Ua1bU69uTWVsE2zlGkzlBHx0npetXIOJlIA1HYMTiQRqMgEtysby174/FlguF/vjYvHj7l18PqOxFxf3nnvu3Yu7X55z7tlULpcLAAAQqw1xVwAAAKRSAAAagFQKAED8pFIAAOInlQIAED+pFACA+EmlAADETyoFACB+UikAAPGTSgEAiJ9UCgBA/FqWuf2N3I3vfO/b098/P33pfAihfdPW9rdtvfcd921IybsAAERVfyr92uTvf/lbv/fVV1/8wZU3b29t3XD7xhDCjSvXr2Szb739jp+//+EP/fQvfaDr51auqgAANK1ULpdb6jbf+POv7xv9375z8dsb77r9vV1/670/9re2vHXL22//0bnrP/zOG9+ZeuO7Zy9895UzE7mLV++9676Rgd953zsfXI2qAwDQNJaWSq/duPZbo//48//l0I2tG//Bgx/9e/cNbwgbrty4cuXGlSvXr1y5cXXhxZWL2Qtffe0rf/Qnf5g7d/Uj7x3+7YFPt2xY7mgBAACa1RJS6aUfXvq7/+ZDf/y9U+/4G3c//d5P3rdpRyGDXrlxdeHFlSvXr2QXlsxe+av/dPY/fPebf/pT79j57//+lze9ZdOqHgwAAAkVNZVeu3HtkWd3T/zVtx5+/y/+2k/++kIT6S1h9MqNK9kbV65cv1q85MqNK39w7j//+bf+pPvtP330o8e0mAIAsFjUJ+V/a/Qf//H3Tv1C34c/9pNPXr1x9Y2rl9+4evmNq28U/Xf5++UWvnH18s4fffc9O3f88fdO/dboP17VgwEAIKEitZV+48+//uF/+fCmv7nlXz70uY2pjcXtoPnhpNlFS0rWyd648vLMH/3gWxe+9KsvevgJAIASkVLpQ//nrj9NTf6T/t9+IP03qobRm336O992/4/env7S6/8xu7Dk+1cvnf7vL7/rxo7j/+jkGhwYAAAJUnuU59cmf/87F7/9t3/2g/e97a+/cfWN4uh5ayq9OZz03emfev/Wh37k9juz1688f/YL2etXrty4cvXGtTs67vrOqW9/bfL3zWMKAECx2qn0y9/6vWvp3EDnh964+kYhemavX7l87Y0//quJmR/OvHnt8g+v/aAl1XL7xrbbN9z+P239hQ9sfX9nujOE8LP39GVvZP+v//7c5euXf3j9zeyN7PVNN778rd+TSgEAKFYjld7I3fjqqy/++E/dmwob37h6Od/qeXHuwv8784d/dunVqzeulKw/3PnY4LaBn3j7O/M//kjr2+7Y2PoXP/jv2RvZ/JJUOvfVV1+8kbvhK0kBACiokUq/871v/+DKm/dt6cr33WevX3n9B69/+ezhN69dXrzycOdjH+36tft+9K/nf8xez7745//h1//brxciaQghd0fqB1fe/M73vn3fX9uxgocBAECi1Uil098/f3tr6523ve37V9+4cv3qn79x5quvH72yqIk0LI6k1+a+8t3/+Ov/7dd/eP2Ht6yXCre3tk5///x8Kh3f8+n3P/dmaWGtobWttb093d33wN59PX2dqz7H6eTIZ971mxdDCK27PjCZebBztfcHkby2r+tff+p0CCE89OyT43vSIYQw9fW+rq8dz4YQ7vr4tz420n3LBnNT09Pt7Z1ta19VAFimGt3o05fOb7h948ZUyxtXL//Fm2dffO3L+Uja3tb+r97zL99++9vzq5WLpC/+w8WRNL/L2zdOXzpfdbfZkL2UPXt65uhzX3t/z6EDE9eWfmCwzsxdGN13qKv7xbHpuGsCAPWI1Ap57cb171/9/h9O/+frueshhPa29n/x7n/xvrt/5kdu+5E9L3904O5HykXSf1g2klbT2roxhBDC9Wy2aOnM6/uGX+rJ9Pdo/oFKpiYG+o4ePRtC691xVwUA6lMjlbZv2nrjyvWL2QsXvn/hYvZ7IYTWDa3/4m9+5m//2EMbN7T87I9/8NDGz2298+7iSPqfvvuV6pH0xpXr7Zu2Llq8493jkw/3zP8wNzU+sW/PSy+cvh5CyJ6aODj+wZ5+31UKIXT2jE51z4UQQki3Lyycfm3ybGw1AoCVUKMHv/1tW69ks6//4LU/nZ3IL8neyP7bs//2le/9SQihZeNtD/7YQyWR9GP/7WPVWklz4Uo22/62xan0Fm2dfT2HD71n4YGo7MTEhUiHA82vJd1+Z3v7ne3td+o/AKCJ1Eil977jvrfefseZ83+a77vPG3396D/9k3/68vn/mgu5wgRP2WtzX52qFUlDSL2Ze+vtd9z7jvsi1K27vat1/uXs9FzRL+YmR18a7jvQmX66re3p9q6DA3sz41O3jD2dHPlMKvVUKvVU596p6czXh7pH2tqeTnd+Zs9ombkDSs1OH973+Z7O+cKH9k1MzC5aZ3rq4N7P93V9Ot32VCr1VFt6pKvn0J4DZ6bmbllrbmrywJ5DPV0j+dXS7Z/u7ju879Bri8sLc7NjB77Y3/3pdNtTbemRzp5Dew5MTs4tXm+xy5lDY0N9n+lMP92Weqqt7enOroMDe74+NrV4MG7t87ZwBl47tPfzPZ0jbW1Pt3ceHBp5ZWpu7lD/U6nUU6nU00Oj+U2ujQ7llzzVPnymaONrY0NP55enhyeXeIzXxobnt+07OBtmpw/v/XxP10i67el052cG9r6cKXfiJkdfGu7/TGf7021tT6XbP93d/8WR0QuLzlzkYy/j2tTYS8N9B9rTT7WlR7r7Rw9m5kLmxc7UU6nUU23dLy0c5Jk97fMnZOBw8f6n9nbOL+8/tOgy7v9MV/vTbamnUm1Pt3d+pm9o7PBErXd96ut9bfkCP7NvIoQwd6j/qdT7vnk6/9vs6x/9iadSqacG/p9Xhxfq0z1y61DTuTOFX/Uc8O89ABpEjV7xDakNP3//w1/8iy/euOOW5X/wP/7gjWvffzr8s/dsfW8qpLLX5r469Z9+7ZvVI2mqdWPrjdm5n7//4UiTlc5NvDa5MMA03X7nwuLZw8OHhp+/VBh6OnN65ujpY0cPTTx5+CMH+u8sLWUyM9R/+vilEELInr2W7ly0wq2ys5N7+saPnbpeKPyFTx0dHT1zeOwXBzoXVpp6eaDvK0eLOkyzl7KnT549ffLs6Nju8bGernz9My/19f/RyUs3V7s08+apmdOnjp8+PPbI+OHuQnlhenLvwBefObkQ/LPZsyfPPnfy7KGD99+y3zJmR4c+O/TCmzcH4mavnz09c/b0zNHDr3xybHjfzdG4kc/b1MRQ39EXFo5u5uzMC7/5pfGxXUNVz1ttSzzGuclXhnq+9sLphZ/PXjz6zFfGxqbHJh7uKxzT3IXDw58ffuHmQWVn3jx17NVTxyYPP/GLYwe7Fvq3l3jN3FqR8X2HBj41s/A2Zk8dO/XR8anM3vaqW9U0N77vswOfunjz6shenzl7cebsxeOjk+Ojew72L7sZ9PZ7hwfueP65N0MIk4dfmdzX3lXYd+aV8ZkQQgitdw8PbF7ujgBgZdROhx/66V+67dKGcKN0+X/9y5c/ceoTL5//L3PzkfTXqkTSlg23ve8dD3W03XPbpQ0f+ulfqrXPucuTY18fGlpo/gl39PSkQwghXJvY9/mFeNG6c/f9Tzyx46EdG0MI4dLMM0NfPDRVWtLZY/ORNIQQdmwf6C5dodTp14+dur5l172PPXH/o7vuyLfVZk+/OjycWSj78qG9Y/ORdMtdjzy288kn7n9k53yj7syxrx8Yz7+8cHDvfCTdtnvnJ39397PPfuDjj961KV+rF8b2HS602l44OLQQ1zZt2v3ozice27Zzy/x+h4ZeurW98dbzNPbSnnwk3XTXo0/+zLPP7n72kzt3bwv5E7J/z/j8sIslnLfpkaGbkXTLzm2PPnrvQ9s2zhw/+czxWqeumiUf48lnvvbC6Y07HtrxxBM7HtqWfwwuZE9/c/+hQnvptcy+m5F0y867H3n03vmDCtdPPTe6Z/4ML/maKTY7OjpUiKSbNj30yI5Hd9+1JVx6/lOnlzWMM/PSnvlIunHbQzueeHLnY4/evW3+art0aGRiKY/Rt3T27XjkoU2b5n9s3bn73kcf3dHX2dI3vD1/LWQnJ0dvnuJrmcNn8pVv7euu+m8eAFhLtZ8g+kDXz917131/+r3J61tSJb96+S9f/sSpTzzxzl/99W89WSWS/shtb/vA1p9v29g6Mfn/bb/rvgpfN3r6m+9LfbPs9pseenBvX0sIIcy+sv/gxWwIIdzx6BcePzyUj6qzhwYO/srRbLh0duTAa8MH7inZfMvuD4we6ukO0+OTbT2htp1P/i/jB7anQwhhLrP3s33PXMyGcOl45kCm50BPCHMXZtvad26bnpxO7x/bs6+7JYQQ5qb2dj//zOkQwpuTk7OhLx1mX5uPhK137z80MJxvWdvT0995aN/End1dm3s653c3NzY+cjwf1+5+NjO8p6slhBCmXxnu+dLzZ0P25MsjY32HKjzpNTn+Wr7Na9vww4cPzJe4Z+ieoaGJua7NXd33tM2F0LaE8zY3Nn7gZL6Yjbs+OTy27550CGHuwqGhz/7K0WyoV13H2PrI5x4/PLy5LYQw+8pQ15demAkhhInMhZCftnN6Yv+h+Uha9JbNHh44+MtHsyFkx0Yyk0Mf7Krrmllw4eDI6fwZDtt2fGF8aKgzhBBmx1/s6//mqfrPR5iavta5a9PsxKUwMJA5/ED+6pg+dKjzV85mQ8hOTk+GELkxtqVv31Bf34td7/vmpRBC6+Y9Bz+ypzP/q/cM7Tj1qdMhZC+Ojl7Yty/fLPra6Fh+huCNfUNdy2zyBYCVE+lrP0cGfue2mVS4klv8q5f/8uV/8PKvVomkP3bHtqGf+Ht333H370+9mHv9ysjA7yytgq077j90aKFPfPyV8Xyz1bbte+bjRQghPbRn+5YQQginxycXtbrdtXfkwZ72lrb2e/r7InRWbrp33/7tC0W39ezvG5hvg7qUyVwIIYS2zr2HH5+Y+t/nZhciaQhzs9fCQo/r3Oy1EEJoa2nLL8m+vrfvs8P7MqPj09NzLX0jj2fGhg4e+OBwz3zHcWZ0ar7hqr9nuGshmbU/sHcov+Ps+OhrlSrblm7Jt6+dPXi4Z2D0wKHJiam50Pmew5nHRw8NjOx9oKttaectM/raQgh7YGTvPfOrtm0eHnnProUxvnWo5xh3PLAvH0lDCOmuob755tK52cv5cZez45PzB9Va/Jalh0Ye/t1nH/nqN56YGv9gV53XzILpM2Pzrc0bd+9/OB9JQwjpvv79A8s4HSF0DgyMZfZOz31i6tADC7lwbnauZb5+c3NlBtDW457h4bvyryZGFw4z88pY/s3Y1Dk8UGNACwCsoUizLb3vnQ9+5L3DX/jjz7+5/Vq0HBtCCBtSG973jgcf3NLXuqH13575N9nT3//Ie4ff984HI2zaunFTuq29c3PfUM/ePV1dC4FvanJ2vi/17Kn3p06V2XDqwuRc6Coekrelvadmr33xnjvbu9NFP6fbe7rCCyfn9x5CIddem52aGhs/Mz4+PTExPXH65hSrcyGfSrfvGbrjhefeDCFcOv368596/flPhdDauqOnc2DoPXuGty98+87lycn5r7bKvvClt7zwpcVVmp66MB06y7ZpdQ29p2fk2PFLIWSzJ4+eOnn0VAhh07a7+vq7hvb0DHXPZ47I5+3y1NR8ZVq7t98yQWxXZ3f6j07OVDhrNdRzjK2dN8dBhtCSTreEMD8mNd/+OzVxYf6cd23uKn7Luh7YW7RlPddM0a+m5veR7u4uDnAtPX2bW194fRmtpSGEEOYuT4ydyYxPjU9MT2Quni26hlZK11D3rv1fO5kN2YlXxqYe7OoMmdEz+YExm/q7+9M1NgeANRR1DtDfHvj0q+f/ZGLqWz/svBElmKZvf/sjP/7hv/62HW/Z+JYXvvvvpv/ku93v+OnfHvh05S12vPsbN+crLe/WJ/HLmZubzWeWmxW5c2mfvG1tt+aTtvTC9vONoCHMTby8Z89Lh0/eTKKtm+7YEt7MDz9smz+nbX0Hhr8Qvrj30MzMzYdxsqePn/7U8dMHRraNjH5kb3dLCHNzNSPIbOVVOnsOj14e3pM5dvrmFAmXzl48+twfHX0uM/LEwNjBB9qXcN6uFfaUTpeeh7Z0CBVS6XwQv/njotLrOMbSN6IlhGxx6XNz8zttbWup8mRQPdfMzV9dm9+4taX91suoLb2E2XPL1eDy+L4v7jlw9vTNYLtxy7aW2bPZbAihrXx16tH5wHDP104eDyE7c3j0wt69c6Oj8y3MA8NdQikAjSTqZ2vLhpZ///e//Hf/zYf++MypNzuvhttLx5gWtG5se2jLzz7U/v47b7tz7toPn5/8V2e++Wr3O3763//9L7dsWN5E+G3tbSG8GUIIO3Z9I/PB7nIf3SVZpjVdLbIslp29fGvn6dzsws/pfBCZndwz8JXnz4YQwpadO4b3PNDfc09Pd8vh/t/5lWP5GhSqsnno4J6B/a+Njb4yOnpmPHPxbOFB7rNn9w2P9098sCu0FCq85bFHJw9uL1fbaofQ3vfBsckHJ8cnRw+/Mjb+WuZmq+31U8+9uLd/++GBtsjn7WYEn52buyWqzV2erfIAzq2zO83NXb/113UdY61k1tY2fzFlp6v1d9dxzdyUbpmP4tlrJam5etgtJOb5nxbVb/LA5/s/NZMNIbTesXu4e2hge0/PPV0TL3a+/9RKz4WfHhjetvf42Wy+E7/v8li+pXTL9qE+X0sBQEOJ3B8fwqa3bDr60WO//FMfaZtMbZzJlTyVvyG1YdudP9F/9y/81t/4J4/8+ODtG1qOn/+D/+P3f+v0H5365Z/6yNGPHtv0lk0VCo6sszM9P5pvanoqtLS1zf83N3lmLDM9NXszqRSpGDnKm5oanyr6cXY6szDqsLN7cwhhduzlw/ngsO3+Q5mhkT0P9HWn24rCa5Fr01OvZaba+vb0Hxr72NTsJ85/69FnH9s0/6T15GuZ2RDCnV0Lk7LOTl6YbSscVJjKTI5PXJidq3EAc7Ozk5nX5rq69x38yPjkvtm/+rVvfGHXQ/khkyGbGZ8OSzhvbZ2d829SdmJqongvmYVBnLdYGOt5S+6anZoqWW25x1hWZ/fm+UKnX5so3mPmxa70SFfPZweGM5m5+q6Zwj7u6Vyo+HimeKbb2bGxC4u672+G79nZolQ6PTtVmmBfO3RwvgH9oZHHxw5+cLi/syvdEuauLZzGpefFm0dROglr+0B3/6YQQshmXhk5cCY/iGHbQHffirXHAsCKWEIqDSG0bGj55x868KVfffEnc11v/bMNt70W3pp9S3tbx0+mH/jQtl/qecffum3DbePTX/u/v/W7nxj9R2Nf/dI7r2770q+++M8/dGC5raR56b6uvvnHY86O7J+a/wifmz6454XB9//rd239Z+mhMhPeL012ZmRPZmEi89mxfS8tdHje1d+XDiHMTi+kk6KoMTcxcbgQ4uZCCGF29HB72z/b+hP/+v19Xzw4mQ8KLe3dXcPD9yyMnsy3Drb0DHTmM2T2ZGZ/YZL/qZf3Dnzp59/33Na3j/QdrHRM0yM9T7/l7c+8633/bmDvK/mV2tKbe4Z6hhYGVuYjV/Tz1j3QmZ9LKJye2HdgoT1wenLv3lcWteG1pNvnc002Mzm+UMfp0fGDpaM3l3OMFaX7uvrmA/7MyN6X52fjn7twaP8rpy9lT598fWzicrpteddMunOgZ36qqfGRsdH51uK5iQOj+49dX7R2W6GXf3LszNT8y7nxkUymNMDOTZfZ5eXRQ1MLyT/i9P637j7/f9lrpf9ASncN97eGEEJ25vnnL4YQQrhjYLhTKAWgwdQTFt/3zgeP/6OTX5v8/S9/6/e++uqLf3nlLy63/o/Xbv92COHGletXstm33n7Hh+//8If+51+qMAlU3dq79w9//dgzl0IIp575fPdE10BP2/T4K/mnkULrpuG9Dyx/qNzMsWM9nS/39dw5N/na8YXxmlsGHhzuyldhczrMzIQQZl4d7m/ZN3xPmJg8eOg7hQGCs7NzIYR033sG2k8/dzaE7Mxv9n12Ymh7X9eds5NTo4fn57nc1P9AXzqEENr6+/buOvObJ6+H8ObzQwcnB7r62ucyo68eX3hyfO9QpWNqHxpuHzn5+qUQzr4w2jN7Zri/PR0uT4y9cvj4/AkZGGpf0nlr63tw70Ov/Mbx6yFcP/4bz3Ue3taTnpvIzJwt01AaunvaNz3znUshhJnTQz2fHerbHKZeGz12cfG6yzjGytq79+/5+vgzl7IhnD36le7Or3d3tc1NXjg1k3/LWvv39XQt5djLSQ/t6xo5/urZELKnXx3seu2hvnSYms6cypZ7zmlzT88d4dSbIYRLx77S13Omv7tleuLM2MnFK2/u6QrPz4QQwvH9h4ame/o7L48fnjh8fP6ZsDB3reJQ14pnI51uzY+8vbi//8Dh9Fznvo+NDuef0GrrH96+5YVXb44K3tE11KP7HoBGU/9n0we6fu4DXT93I3fjO9/79vT3z09fOh9CaN+0tf1tW+99x32Rvr1p6Vp6RoZ+d/LQbxzLhnD97PFXi6Z2b919YGhk2Z+12x7b1T/x8nOnLh47erGwcNOudx8+2J1v42zr69u/+8xHj2VDCDPHT/3G8XzD4MZt21rOns2GEKYnZ2fDPen09gOjH5jq/9qxmRBmZl54ZuaFor207rz/0EKBIbTvO/zwRP7rlLJvnnzhmydv7njLxw8PDFQOTZ17fvHQxOeHnruYDddPHzv1m8eKWylbd48M7Z+fuyr6edu899DDmYXvdpo5efZo/rQ8cm/72HdO3hqv0gN9+3ZN/ebJ6yGE7OnXnz/9egihddu9nxyY2/9MyfPp9R9jZS09Ix85OP35PS9cyoaQnbl0svAFTGHjzid/8eDQnUs89jLS/Q8f/viF/vxE+pcuHT96KYQQNt31SM+1o8dK4ndL374Hd48eOzYTQghnT55+7mQIIWzZ/TPDIfOpW9pW00P7dx7oP3U6G8KlSy986tj8tbHljm2zb57NhpCdnZgMQ0uZOyK0d/Z3hfzFmJ25dGomhOnLIczPG9DW1z2w7dXnFpq7dw68J8rEvQCwtpabHTekNtz313b0bu/7O+/+5b/z7l/u3d5331/bsUqRNIQQQlv73rGPfeNzux59aNO2TSGEsGnLpl2PvPtz3/jY2J72FeiUTHcdzDx+5OM7du1obW3duGXHlkc//uHM+MN9N2PT5j2jjx/5+L27trW2tobWTa07Htrx8S88PnH4gR0hhBAujU/ku7Pbuh8cm/xfv/DJ+3fv2rRlUwghtG5q3bFr2xO/++hE5hcHiqdB6uw+PPHEFz55/+6dd2xqDSFs3LLtrt1PfOAPJx4f6al+TOmBg3smv7r7yUe37Ny2sbU1hLBx07a7Hnp01+e+8bGxvUUnJPp56+w+nHns2Sfu3rllY2vrxi07tz3x7GOZw+8pM3dS2z37xoY/92T+VOTP1S9kJj4yXDZO1X+MlbVtHj68Z+LIzzy2+65tWza2tuYPaufvfnVP5sD2myd4WddMW8/I45kv7Hp01x2bWkPrljt2PbrrSObxkf62MhOWdvaMjn/4k49u2bFlY2tr67Zd9z757GMTY309izJ3um9gfHz3E7vv2rIphNb8qdj9hxN79vfnBwxcHBtbyrc7hRBC+97DH35y911bNoXW1tZtO+/u6SxK223bhwcWvjO49a6hYXPnA9CAUrlcmbnxocTkcPqF5y+FEDY+emTf4YH13v87eeBg92/MZENo3fkzExMf7Kq9RbzmRocODL6QDSG07vqZiUzjVxiAdWj1GjWBxjA7/tL+0fkhFd1D3SIpAA1pvbd5QfMaH+sfmQ5zlycyF+dnotp0797hCN+7CwAxkEqhWbVdmzp29vTNn1t3j/TXMdsBAKwJPfjQrDrv6d5xx6bCM3lHHh/do6EUgIblaScAAOKnrRQAgPhJpQAAxE8qBQAgflIpAADxk0oBAIifVAoAQPykUgAA4ieVAgAQP6kUAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIn1QKAED8pFIAAOInlQIAED+pFACA+EmlAADETyoFACB+UikAAPGTSqEhpFKpuKuwXhROddlzHvsbkUqlYq/DclQ/vUSx/FOX9KuIdUsqheVKLVip0kpesN7kcrnC61SRGKtUYnFlVqN6K1JmxLOXlPMcUfElBAnSEncFINlSqVThA6D4dd18nFBQckWtyAWWICtysPlCqse7/G+jrLnG1tXbDUEqheUoSQm5XK6wpEqeKP7YW/ypk18zv07hwzJKOikUW7JmlYXFy4v3u/jjuWaZZWtSc8Nl7rTumlRXsw5LqtuS6lw9dy7pTayyfskm1cupdIlWquRila7ekl1XujyiH9dyMlxh28V/btX3FaVuS7owSva4pDMQ8UCgYUmlsKYitn4t9ROlbJNtSdoru7BSW2+l1Wq2DVePINULWdJO665JFNXrULMadZ+9iKr/m6dmfYo3LFvPZVYvuurVLlbzGljtOkf/w6njr6/SG1RSgejvLCSUcaWwpur71CxuJqn50Vv82+KGk+oLq/xYaXeV8nTNipWtzIrvtO6AEqWeNetW34EsXidVJMq2NXcX8YQ3Whvbyr7FixX/fdWsQ8RzVd9f31IvmJobQoJoK4W1thpNGpV695ajbD2j7GjFD7DsTte+Jks6IXWfvcWbFJdZ0q68fEstp9E6hVfkPBTSf7j1PVqNfQFVSKWwphb3OUbcMFerT3/F+15rts1E7Ddffk0q7XQta1KltOp1K7FKPfj1lRMS3sC2xuch0ecKEkEPPtSvpFmlUs5Ywaa25wLCAAAgAElEQVSXunexeDRhzQrUsUn1WkUppHqFqxS+pJpEXDnKG7ekdWpWY/EVFb22EVer9KZEL2o5Vm8XdTT6Vnodpel9SW9NlIuhvj+x6gWK0SSOtlJYlkp9siXLa74uq/hRhurrV+njLltIzX7kSqvV7IOufrAllVnOTuuuSRQ1d1fpdR0HUkVJtWtWpno5i1eLeDEU1Bd0lvNGLL/kktAW5VKpsq/o52pJf32V3qCa6t4QGpN/S8HKWJuWCe0fTa96cF+pXdRXvsuvwaXKPZsPCeIWAytmtT+zZQLqJq80K+8szcSHHCSAzxuWb0mDB0gQ7yxNQyoFACB+nsEHACB+UikAAPGTSgEAiJ9UCgBA/KRSAADiJ5UCABC/dfSNo6lU6si5I1HW3Bw2RyzzQrgQcU1lrkaZvR291Vc7ce5EvszBjsHqax45d0SZylSmMiuVGUWy7p/KXG9lDmwdiLhmjNZRKqUpzc5W/FU6bU1rrt81S/4RPtgxWOmf5SfOnSj+sbejt2SJMoG1oQcfAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIn2fwAZrQ4tmRSpZEnCkPYM1IpQDNKfpsUwCNQA8+AADxS3ZbaSqVCiHkcrmSJXnFywEAaGRN2Faay+XkUQCAZElwKk2lUiXps3hJLpcrbjcFAKCRJbUHf3EkZR3KhDBZ+aGNTAhXI695Xplxl0lcejt6qy8pfDV89Of64y1zNSyuZ4kq9SyxqvWEREtqKoUQQk+tp4xPRF7zSOQ1lblKZRKj6E/rJ2XNFVcykdZgx2ClqbVKQmdvR68YChElsgdfQykAQJNJaltp8ZhRIRUAIOkSmUpLpoIqecIp/6OoCgCQIIlMpdV59B4AIHESn0pLGkS1jwIAJFHiUykAJFH02aZgnZBKASAG0WebgnUikTNDAQDQZKRSAADiJ5UCABA/qRQAgPh52oll2VrrGdLzC4P3b+vorb7m1YWvio5eJs0kE8Jk5a8yz4RwPvKaV5W5lDKbz+Jn20uWFB4q6l10XypZUvgK++hlrobF9SxRqCckmlTKsvSEMDtb8bfpdDgSec0TkdeUSZtSvNfSei6zKVU/9tVec8WVhM7ejl4xlKakBx8AgPhJpQAAxE8qBQAgflIpAADx87QTwLoW/bl+gFUllQKsa9Gf6wdYVXrwAQCIn1QKAED8ktqDn0qlCq9zuVzN5QAANLJEptJUKlWSRPM/VloOAECDS2QPvqwJANBkEtlWCgXRv356tb+oGppezTmkzkdes4lnm+rt6K2+xFfYQyUJTqWFIaSFptNcLmdc6bpScnPv7eitdLs/cu5I8Y+DHYMlS4Caas4hdSTyms2dy6ofO1BJglNpIXQaVwoAkHSJHFcKAECTSWQqLe6mBwCgCSSyB7/S+FHjSgEAEiqRqTRUTpyS6PqW6e0oXXTk3JJL8Vw/kFCDHYPVVyg86Ll4roAS5gpg7SU1lUKJ3o7bQuhZfBvt7dhabvXM4KL8euLc1eC5fiDJot+Xot/rYM1IpTSD3o7bTpy7urihNIRw5Nz5xQvL3qkr5ddyxS4t6QIANUmlJF4+ki6/nLKFlG0/uBAuLF6zatLNVO/r7+2oZ6QBADQTqRRWUT7p9nb01phR/NyJmvl1drZnFSoIAI1CKiXZVqqhNHb5kQaDHYOV8ms6nQkhFA0SyKTToioAzUMqXVO31Xrm8epCZ/HWWs9Rnl/oLI63zLA6T6xHXDNBkXQljr0nhHDk3JHNYXNYeH42H1XLKh4VYK4AGlYmhMnK12cmhPOR10zGvaDI4uflS5Y0yAOUnutnzUilayr6d0OvxvdNr0aZq/G8Z8QyExRJV+O5/oUSrlYpszAqoFLFyj4KBmtpNe5LCVL9iBqE5/pZM1IpiZSgSBqjmvMPDJZOO2BUAACxkUpJnsGOrSLpiijE1sKogNnZqKMCAGBlSaUkzGDH1iPnzoewOe6KNK1KbaXmCgBgVUmlJMlCJCU25goAYJVIpcBKmp3tSadvmSvAqAAAopBKSQwNpWtglR77NSoAlmPxjEslSzzqTnOQSkkGkXQNxDWri1EBUFMi5pCCZZJKWVmZouhQY83eRWtW6skVSdc5owIA1oOkptJUKlV4ncvlyv6qZDmrbbBjawg9i2dX3lzuefmqs76XyIikLLbMUQF5gx3BLGMADSKRqTSVShUnzuIfK70mESrN+h5CQ3ztHglSc1RAyH9v0LkjEfOr9leA1ZbIVFopa5bEUJF0LeU72SN330OjyLeV5kcFVKL9FWANJDKVVqcHf+0Z97kORX/AYjXWjMVqtL+GIOmuvEwIk5XPZyaE85HXbOKTHu9z/YMdg9VXKAwGW1zPEoV6Ri+ThpXgVFo2fRaaS1OplB58WCUlN/fBjsFKt/voz/VHL7PxRW9/vRAuhBVKuiYlKOip9cT6kchrNvd8SzE+1+8eQlkJTqVlx48Wvyh+IorVo6EU1kDNpJuflKCk/VVUBRIkwamURiCSQoPIz59V0v66eP4sORVoWIlMpZW65vPto4Ue/DWvF0BjWZxBCznVWFWg0SQylZb0zpuvNC4aSiFx8jm17FhVzahAvBKZSkPlxCmJrhmRFJKuZKxqSXe/kEre4mfbS5Yk7qGi6M/1s8aSmkoBWFklMbQQUn2DADE+rb8aoj/XzxqTSqlHb8dtGkqhuRX6+hd/g4BmVGA1SKXUIXPi3NULcVcCWDMl3yCgrx9YDVIpAEtTqa/fc/3AckilccrfypPVzJBOZ0LoCU3+jSfAEniuH1gRUmnMZmd7EpRN0+lMfqZugLI81w/UTSpda0WRLhNCPuHlmxny9+6r5daMXuaqrymS0gga868juWuunqZ5rj8TwmTl85kJ4XzkNa9GXrOOMqPP4rQa9bzt1hmXMiGEW5dcXXjUfeuttcqEEG5dcn6hnvGWufh8liicz+izTcVbZoOTStdU8dwTvR0lS/INDLflv8O65AIa7BisdElFn+FimWX2dtx24txVfffEbs2u+XVS5lpK7nP9PbVmRzoSec0Tkdeso8wQeRaneOuZlDLDUmbFSsqajUwqbSwnzl1d+A7rfDZtlNmXFiIpwMqo8lx/gydUYJVIpfHIh7x8c2lZC/frlcmmg/NtEiUyg4sqIHoCsShOoreORm2Uf5wDq00qbWhF2bRMgqwgszjsls21ZXv0eivkV2kVWDOFhJpOh8GOTPWVgaYhlSbAkXPnyybIzWHz4pXLjjOLPuN92fTZ29FrOCkQiyPnzufvdb0dmcLwOF380JSkUgASoafwPEehi188ZVVFn3+AFSGVxsCTQwDLUdTFX9y/777KCov+XD8rQioFIKmKR6D2dmTyM+sBCZXUVJpKpQqvc7lc2d8uXg5AsyqaWc8IVEikRKbSVCpVnDhLfmxwuu8BVlmZEahBQoWGl8hUWj2D5kNqcWMqAOtT2WlQBzvMzQyNKJGptIqa7aar8f2zW2uVeb4BvugPoJLoX0gY75rLVPi+0yPnjiTr+05pfDWf1r8aec3zkdeso8wGl+BUWmgNXWr3/Yp/q2z0Z/R03wONpmSG47JzHueVzJpcdh7l1StzZeVvxb0dvb7vlBVRMwmciLzmkchr1lFmg0twKi2E0UL7aLIGmALQICp836l4Cmsqwam0rOLhpEIqAEtVPBlqb0c4ci7e6sA6kshUWiluNviD+brvARKl58S5E4URqJpOYbUlMpWWPGLfaOkTgKZx5Nz5EMJgx6CvOYXVlshUGiIkUVEVgBW0+GtOxVNYWUlNpYmj+x6gOSyOpyGEBpl7p2lmCGJ9kkoBoB5F8TQMdmSqr7w2mmaGINYnqRQAluvIufObw+YQQm9HpjChtS5+WBKpFABWUE+htdIIVFgSqXQtDHZsNagUYL2pMALVxwGUJ5UCwOoqHoHa25EJYWu89YHGJJVWVPNJRv/aBVhL6cr35DVYc6WcOHf1QrgQQhg0AhVuJZVWVPNJxhNRS8ocOXc+hM0rUiuA9enEuVtuur0dvSVLCo6cu+VB88GOwZIldZS5OoxAhVtIpQAQs0aeAxXWjFQKAI2iAedAhTUjla6udDoTQk8wbzEAS2QOVNYbqRQAGlzpCFTZlKYklQJAbWv/tH5Z+TyaTmcEU5qPVLqK8neNBrmRAVC36M/1r43Z2R7BlOYjlQIAZQx25Gf7z1RqXhGLWVlSKQAkz8o2l946I1UIIfR2hCPnzocQBjsGK83end9qsKOwIJNOi6rUL6mpNJVKFV7ncrmay9eevhUAVlU+mIaw3M+axR9Y6XQ4ce7EhUgVCEfOHVmYK6B3drZMwA1aVYkmkak0lUqVJNH8j8XLU6lUyWoA0GQWgmn934G94m0oZUvLR9XejhDCzVEBoiolEplKK2XN4uW5XK643RQAKLFm3Xr5VtUT505cCBfCwqgAraqUSGQqTa7ejt7qS9b2K5gBWHnRp15ZoTV7ejsyR85FLaqozEwIkSaKWbUjqtiqWjJWNeIohdWoJ2spwam00BRatuk03u77Sv/6FDoBmlvJfb63o7fSnT/6bFMRyrza27E1/3BSZJkT566GcGJt6xmlzPmjyI9VXSiqTKvqiXO3jFtYjXqyxhKcSouHkJYEUCNKAVhXjpw7vzCRU229HbeVRLoGV6htcYLs7bht4fc3D3yJ0ZzGkuBUWolICsA6dOTc+cGO2o/kJy6SVlI4ivxY1byiaJ4pDAMQVZMikam0Su5shEhqTigAGtNgx9bmiKSVFAJocQ9+2Vbk5j4PCZXIVFryfH1xV36Ib8rSss8SAsAaqj61fubIufOhaLzmOlGIqreOVS0MAMj0dizeZA3qRalEptJQIW6uWQY17xoAjanSdz4tzLdf/gGgdajsWNWC3lrftrr8Ly9gsaSm0jVW0g5adt41AGgEi4Np/kfTIUVXdijqraMCfNvqypNKa8jnUdcZAAnlaYfVUDIqoMq3reb1dhgVUJtUWoV/9wCQSIXmUpF0LVU61fn+1QijAkIIYbBj/T6JJZWWsTACuke/PAAJJZI2mnzzavVRf+l0OHLuSMT82nztr1JpiUzh3yiLn8gDgAQRSRNqIYf0Vs+vVdpfE/rWr7tUmglhsty/PHpCJoSQCT3nzx0pzJqRiG/U7e3orb6k8GjhYMdgyZolSwqDuJW5smUC0PRWIzPUXLPQ/lr3LhrKukul5xd9T24ImbDwvp4vmjJjNb7LeDVE//be6LVS5sqWCUBzW43MkJQcsoLWXSpdJOOLyAAAYrch7grELpEDLwAAmoxUCgBA/KRSAADiJ5UCABA/qRQAgPhJpQAAxE8qBQAgfkmdrzSVShVe53K5mssBAGhkiUylqVSqJIkW/5h/XRxPAQBocInswa/UCFocT3O5nGAKAJAUiWwrXQ8GOwarL0noV9yuT70dvdWXFL7aOPr7vhplAk0gnY67BlCvBKfSQlNoU44fFRSaSSEg1hT9fV+NMoGkK/l7H+wYdAcgQRKcSgthtGRcKQAAiZPIcaUAADSZRKbSSo8xFT/hpAEVACBBEtmDX/J8fUn69Og9AEDiJDKVhspPOGkfhbpFf64fAFZcUlMpsOKETgBilMhxpQAANBmpFACA+EmlAADETyoFACB+UikAAPHzDD6wZIMdg9WX+OptAJZKKgWWTOgEYMXpwQcAIH5SKQAA8ZNKAQCIn1QKAED8PO0ErKLejt7qS06cO7GG1QGgcUmlwCoSOgGISA8+AADxS2pbaSqVKrzO5XI1lwMA0MgSmUpTqVQhcaZSqcKPxcsX/wgAQMNKZA9+cdaUOwEAmkAi20rrtutStIW3ld/83h8sWnI1hBA63lXjO8HP/dmRSmWW2cvVqGsqU5nNVGb0v6OO7TWe6z935kSlei7/712ZylSmMhNZ5tbyazaUxKfS4m76XC4Xy7jSc3/mO8FhuaL/HZ0747l+gCaU7FRafSCpcaUAAEmRyHGleUInAEDTSGoqFUkBAJpJInvw84NHFw8hjWtcKQAAy5TIVFolbkqiAABJlMhUChDF4tmmSpaYQAOgcUilQNMSOgESJKlPOwEA0EykUgAA4ieVAgAQP6kUAID4edqpoo7tvdWX+DJuaBrR/96jP9cfb5kAiSOVVuTmDutH9L/36M/1x1smQOLowQcAIH5SKQAA8ZNKAQCIn1QKAED8PO0E0ISiP9cP0CCkUoAmJHQCiaMHHwCA+EmlAADEL6k9+KlUqvA6l8uV/VXJcgAAGlYiU2kqlSokzlQqVfJj2dcAADSyRPbgF2fN4tclMVQkBQBIikS2lVZXvQffbCkAxTq291Zfcu7Mifnlke+fq1Em0PQSn0pL2kcLP5b07Be4wQEUKwTE2mtGvn+uRplA00tkD37B4txZ+FH3PQBAgiQ4lXqYCQCgaSQ1lZaNpLlcrjCotHjqKAAAGlwix5XmE2elKUvNVwoAkDiJTKVV4qYkCtCUoj/XDyRUIlMpAOuN0AlNL6njSgEAaCZSKQAA8ZNKAQCIn1QKAED8PO0EQFPpeNdg9SW+5hQak1QKQFMROiGh9OADABA/qRQAgPhJpQAAxE8qBQAgflIpAADx8ww+AOtUx/be6kvOnTmxhtWB9U4qBWCdEjqhoejBBwAgfkltK02lUoXXuVyu7G8XLwcAoDElsq00lUrlFoRbEyoAAEmUyFRa3AhatqFUKykAQLIktQe/EpEUgBXX8a7B6kvO/dmR+eUJea4/ej1X49jjLZOGlfhUKoYCsNqiB5oGCZ01Ra/nahx7vGXSsBLZg19QEkklVACAhEpwW2nZDFr85JOQCgCQFElNpWUTp3ZTAICESmQqzTeIVp+yFACABElkKo2SQeVUABqZZ8ahRCJTKQAkndAJJZL9DD4AAM1BKgUAIH5SKQAA8ZNKAQCIn1QKAED8PIMPAA2tY3tv9SXRvyyesJTzGX32rtUocx2SSgGgoQmdKyv6+YweEFejzHVIDz4AAPGTSgEAiJ9UCgBA/KRSAADiJ5UCABA/qRQAgPhJpQAAxC+p85WmUqnC61wuV3M5AACNLJFtpalUKrcgFCXRSssBAGhwiUylxY2gUV4DANDgktqDDwCU8B3rzaRje2/1Jc33VbSJT6X5XvvoywGgWQmdzaT5QmdNiezBLxBJAQCaQ4JTqUgKANA0kppKRVIAgGaSyHGl+SmfFk9NWmk5AAANLpGptFLWlEEBABIqqT34AAA0E6kUAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIXyJnhgIAIK/jXYPVl5z7syNrWJ36SaUAAAmWlNBZkx58AADiJ5UCABA/qRQAgPhJpQAAxE8qBQAgflIpAADxk0oBAIifVAoAQPySOot+KpUqvM7lcjWXAwDQyBLZVppKpXILwq1JNIRQWA4AQFIkMpUWh86ShtLCj7lcriStAgDQsJLag1+nyWirbY5c4IXIaypTmcpUpjKVqUxlxlVmV+Q145PIttJixe2jAAAkVLJTqUgKANAcEpxKRVIAgKaR1FRaNpIWP+EkswIAJEgin3bKR88oU5YCAJAIiUylVRpBtY8CACRRUnvwAQBoJlIpAADxk0oBAIifVAoAQPykUgAA4ieVAgAQP6kUAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIn1QKAED8pFIAAOInlQIAED+pFACA+LXEXYF6pFKp/ItcLlfpV2V/CwBAY0pkKs3HzeIAmpdKpYqTaMmPAAA0LD34AADETyoFACB+iezBrySXyxlXCgCQRE2VSo0rBQBIKD34AADETyoFACB+iezBLwwezb8odNMbVwoAkFCJTKVV4qYkCgCQRHrwAQCIn1QKAED8pFIAAOInlQIAED+pFACA+EmlAADETyoFACB+UikAAPGTSgEAiJ9UCgBA/KRSAADiJ5UCABA/qRQAgPhJpQAAxE8qBQAgflIpAADxa4m7AvVIpVL5F7lcbqm/BQCgASUylebjZiF9FkulUoUwWvwaAIBG1lQ9+CUxVCQFAEiKRLaVVqcHHwAgcZqqrTQsNJdW6eIHAKABNVsqLbSPaigFAEiQZkulAAAkUVOl0lwuV+i1130PAJAgiXzaqSR6lnTWe9oJACBxEplKq8RNSRQAIImaqgcfAICEkkoBAIifVAoAQPykUgAA4ieVAgAQP6kUAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIn1QKAED8pFIAAOInlQIAED+pFACA+EmlAADEryXuCtQjlUrlX+RyuSorVPotAACNJpFtpblcTuIEAGgmiUyl1aVSKZkVACBZmi2ViqQAAEnUbKkUAIAkaqpUqqEUACChEvkMfhWFx/ODkAoAkBxNlUqLM6hICgCQIIlMpYUGUfOSAgA0h0Sm0igxVFQFAEiQpnraCQCAhJJKAQCIn1QKAED8pFIAAOInlQIAED+pFACA+EmlAADETyoFACB+UikAAPGTSgEAiJ9UCgBA/KRSAADiJ5UCABA/qRQAgPhJpQAAxK8l7grUI5VK5V/kcrlKvyr7WwAAGlMi20pzuVzZxJlKpXILwq0JFQCARpbIVFpJcVTVUAoAkCBNlUoBAEiopk2l+d78uGsBAEAkzZlKRVIAgGRpwlQqkgIAJE6zpVKRFAAgiZI9X2n+RSGG5n80ZSkAQOIkMpVWypoyKABAQjVbDz4AAEkklQIAED+pFACA+EmlAADETyoFACB+UikAAPGTSgEAiJ9UCgBA/KRSAADiJ5UCABA/qRQAgPhJpQAAxE8qBQAgflIpAADxk0oBAIifVAoAQPxa4q5APVKpVP5FLper9KuyvwUAoDElsq00l8tVSZzVfwsAQANKZCqtJJVKFfJoLpcrbjcFAKCRNVUqBQAgoaRSAADiJ5UCABA/qRQAgPg1VSotfsKp+MknAAAaXLLnK82/KEmfHr0HAEicRKbS6pOVrmVNAABYEU3Vgw8AQEJJpQAAxE8qBQAgflIpAADxk0oBAIifVAoAQPykUgAA4ieVAgAQP6kUAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIn1QKAED8pFIAAOLXEncF6pRKpQqvc7lcjDUBAGD5EplKU6lUcRIt+REAgMTRgw8AQPykUgAA4pfIHvxcLmdcKQBAM0lkKjWuFACgyejBBwAgflIpAADxS2QPvnGlAABNJpGpNEiiAADNRQ8+AADxk0oBAIifVAoAQPykUgAA4ieVAgAQP6kUAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIn1QKAED8pFIAAOInlQIAED+pFACA+EmlAADEryXuCtQvlUrlX+RyuXhrAgDAMiU1laZSqUIYLX4NAEASJbIHvySGiqQAAEmX1LbSoAcfAKCJJLKtNCw0l+bzaCGeAgCQUElNpYX2UQ2lAABNIKmpFACAZpLIVJrL5Qq99rrvAQCagKedAACIX1JTqSQKANBMEtmDDwBAk5FKAQCIn1QKAED8pFIAAOInlQIAED+pFACA+EmlAADETyoFACB+UikAAPGTSgEAiJ9UCgBA/KRSAADiJ5UCABA/qRQAgPhJpQAAxE8qBQAgfslOpalUKpVKxV0LAACWK9mpFACA5pDgVJpKpXK5XNy1AABgBSQ1lYqkAADNJKmpFACAZpLIVKqhFACgybTEXYE6FT96L6QCACRdIlNpcQYVSQEAmkAie/ABAGgyiU+lGkoBAJpA4lMpAABNQCoFACB+UikAAPGTSgEAiJ9UCgBA/KRSAADiJ5UCABA/qRQAgPhJpQAAxE8qBQAgflIpAADxk0oBAIifVAoAQPykUgAA4ieVAgAQv5a4K1CnVCpVeJ3L5WKsCQAAy5fIttJUKpVbEG5NqAAAJFEiU2lx46iGUgCAJpDIVAoAQJNJfCrN9+bHXQsAAJYl2alUJAUAaA4JTqUiKQBA00hqKhVJAQCaSSLnK81PBWXKUgCAppHIVCqDAgA0maT24AMA0EykUgAA4ieVAgAQP6kUAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIn1QKAED8pFIAAOInlQIAED+pFACA+EmlAADETyoFACB+UikAAPFribsCdUqlUoXXuVwuxpoAALB8CW4rzeVy8igAQHNIZCpNpVKFPJrL5YrbTQEASKJUEpsbi1Pp4h+rbLWalQIAaH6rFx2TOq60DknM3wAA60Qie/ABAGgyUikAAPFLZCotfsIp4qBSAAAaWYLHlXp6CQCgaazThsZCol3qw/vV11/q3P759ausWZK8a5YZ5bgWp/mIK6/IsVepYc0SlrNtI+y95prVr4ey2y7pCqm09yj1r773Zf4d1SwnyrFHqcbiM7zU966+bWPce80rpMpVV2nbFbkv1f03u6Srrsr6S7pb1nHVVVkz+t2mjm0bYe8113SvW+17XXIluK10OfLvaJTW1uIRAqlUqsqAgfrmq4pY1ShKqlppw0p/JNULrF5m8S2m+lmttE6U8xxl2+r1XE4JVbat8ttKe6l+LS1p2+ifzVWOsXr9K21b39/R4sOpft1W33vx8prVqFLyct6RSjVfvRKiX3VheR9gEW8jVVR5f2vWv+y20a+6KnuPcressvfidSrtutKa0e911betUvNlllBl27D0e131ei5p2yX9O6RSPWve68qus6R7XaX1o9zrquy9ZLVmlchxpWtp8T00yppRrFRsLVtafUllOYpvHHX8zdRR+brXX3ERj736MVZ/L5ZzfqpvVXLTL1v/5Z/hiP+WK7taxMaJmqstXif6Wa2+bRTL+VurHh3q+4tbfq3qKL/sR3Kl+td3T1vq3iuVXMc6EWuy1IOqI5ZVKWGZ2ybiqgsV6hnlqqu07cruPVR+K5f6d9F81mlbaeyWdDsrvI7yuRtxTRrKMj82Cq8T+r6vwXW74qEwWXsvvF5qC1aVbcsur15Ifcmy7m2rlxClZHfUFedel3+R0PqvAal0CZZ0E1+Ra67kY8MGzu0AAAR1SURBVKB6BQq/jdgXWXOdkn+rRWxMWn7nQn23reWc+SUdadldL/XYVyqmLOkKCQ2ZDyJet1X2viKNB8161S31Comy7arel1Zq2+olRCm5+jr1NZRG/1WVFZZ01S1uoqvjrr54Sc2rru5ru9K2dV910Xe9/G2rl7DUq67Kb5uYVBpVxKuhUidRHUUtLjnix89y+laKldSzSrXze0wtjIlZzt4X77Swi+obRjnz0Xdacl+oUmAdxx7xxFbPcJVqUn3X0c/S4r0v5wxXKaHsdbuye1/qZ3/x3qtvu5y/9yhXXaVt6/uLy0WYU6/SNV9pLyt7XyrZ+/LvaZVKiHLVrfgdtUQT3+uq7KXmVRfl9lh970v96yje+/I/y6qUUPOqW8FP0uSSSiNZ8X+gFF9tsfzrZ8V3uiKlLa5VvP8ujLj3JVWy7Jkvez2UXW3NTkgiznyIfE6q/MXVvOqW/9e6pBIi7n35b1DEq24tJWXv0S+DiP8mKbv35f+9Ry8h4t6TctUtqcAV/9BZpb3HEhXWnqedaqvjY6+6XJEQbVRTzfLL/ttrzaxIyF7OhnVstVKWdOyVbvRRrococTZiPUv2Hr39rG4R976kbaOrcoajZMSy267G33t9e69+CJWukCi1qrTt8u9LUa665d/TIu59SduuiPVwrytrOX8L9d3rSv4ht6R73cr+E3Sp97p1kkEXW6+Hfes1EfGevqSVI57YKDkmYplLWjN69Za69+iVDFU/46snsLAoW0Sp53JKiLLtko697PrRc2HZU1ffFRKlhLLrRP87Wo29h2X3YC7pr3vx7pb/976kEiptG2XD6nupchqX864tv4SkX3U1/yVZpQJRbgX13ecjHnv1bZNy1ZWstqT3vdKu137vYT2F1PVynAAANDI9+AAAxE8qBQAgflIpAADxk0oBAIifVAoAQPykUgAA4ieVAgAQP6kUAID4SaUAAMRPKgVoTlW+brvuL3kv3hZgZUmlQIKlblV3ISUvyv64zMIbSt3fNe1LqoHV0xJ3BQCWpTgnpVKpOmJTpU1WJIGJcQARSaVAcypupCxJriUL81k2vzz/v/lfFWfcsqUVbxgqBNBCIdFXLqlD9aOosrysxcG90qEVr1O8Sb6G0jaw4qRSoAmVBLuSaFhYJ2LYqlRa9QLLllNz5cVRuPpRhFuDbygXZ2sqe2hVlgOsBuNKgWQrHle6OFGFFeqgr1RafYVEX6fsfquUU/Ooo29SR1EAy6GtFEi2pebC6n3ojabKc/RrXBOA1SaVAutL9A73RlCzlz9IqECz0IMPNKey8z2t+OxRq61m5SvNZlWpnsWtxdU3qbK88dM8kETaSoEmlM9eizvrozwFv/hXlUpbbRGPouR1Hbm50qEla7QDkHT+yQuwvpRN3ovXiZ7aAVaEtlKA9WU5mVIeBVaPcaUAAMRPDz4AAPHTVgoAQPykUgAA4ieVAgAQP6kUAID4SaUAAMRPKgUAIH5SKQAA8ZNKAQCIn1QKAED8pFIAAOInlQIAEL//H1aTDi3fGrFvAAAAAElFTkSuQmCC"
    }
   },
   "cell_type": "markdown",
   "id": "499b7b2c",
   "metadata": {},
   "source": [
    "\n",
    "![Capture%20d%E2%80%99%C3%A9cran%20du%202023-11-09%2015-44-16.png](attachment:Capture%20d%E2%80%99%C3%A9cran%20du%202023-11-09%2015-44-16.png)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33263beb",
   "metadata": {},
   "source": [
    "Nous pouvons voir grâce à FastQC que la qualité de nos séquence est plutôt bonne, malgré une légère diminution de la qualité en début et fin des reads. Il ne semble pas opportun de modifier nos reads."
   ]
  },
  {
   "cell_type": "raw",
   "id": "ab1e0f59",
   "metadata": {},
   "source": [
    "f = open(\"read_positions_5_11.txt\", \"w\")\n",
    "for i in range(len(align)):\n",
    "    f.write(str(align[i]['alignement'][0][0])+\"\\n\")\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
